---
title: "README"
author: "Shila"
date: "15/10/2020"
output:
  html_document: default
  word_document: default
---

<center> 
#### The Spatial Mouse Atlas is a shiny web app-based resource that can be used to interactively browse and download the data from: 
#### <a href="https://www.biorxiv.org/content/10.1101/2020.11.20.391896v1" target="_blank"> Lohoff\*, T., Ghazanfar\*, S., <i>et al</i>. Highly multiplexed spatially resolved gene expression profiling of mouse organogenesis. <i>bioRxiv</i> (2020)</a> 
</center> 

<br>

#### Abstract
Transcriptional and epigenetic profiling of single-cells has advanced our knowledge of the molecular bases of gastrulation and early organogenesis. However, current approaches rely on dissociating cells from tissues, thereby losing the crucial spatial context that is necessary for understanding cell and tissue interactions during development. Here, we apply an image-based single-cell transcriptomics method, seqFISH, to simultaneously and precisely detect mRNA molecules for 387 selected target genes in 8-12 somite stage mouse embryo tissue sections. By integrating spatial context and highly multiplexed transcriptional measurements with two single- cell transcriptome atlases we accurately characterize cell types across the embryo and demonstrate how spatially-resolved expression of genes not profiled by seqFISH can be imputed. We use this high-resolution spatial map to characterize fundamental steps in the patterning of the midbrain-hindbrain boundary and the developing gut tube. Our spatial atlas uncovers axes of resolution that are not apparent from single-cell RNA sequencing data â€“ for example, in the gut tube we observe early dorsal-ventral separation of esophageal and tracheal progenitor populations. In sum, by computationally integrating high-resolution spatially-resolved gene expression maps with single-cell genomics data, we provide a powerful new approach for studying how and when cell fate decisions are made during early mammalian development.

<br>

#### Experimental Design
Using the <a href="https://dx.doi.org/10.1038%2Fnmeth.2892" target="_blank">seqFISH platform</a> along with a novel approach for retaining cell membrane segmentation, we detected individual mRNA molecules for **351 barcoded and 36 non-barcoded genes** in three sagittal sections of 8-12 somite-stage embryos (Embryonic days (E)8.5-E8.75).

<center> <img src="Figure1.png" width="75%" /> </center> 

<br>

#### Links
- The data is available to download **<a href="https://content.cruk.cam.ac.uk/jmlab/SpatialMouseAtlas2020" target="_blank">here</a>**.
- **<a href="https://github.com/MarioniLab/SpatialMouseAtlas2020" target="_blank">Github repository</a>** containing scripts for analysis
- **<a href="https://github.com/MarioniLab/" target="_blank"> Marioni Github </a>** for many of the packages used in the low level processing of this dataset
- **<a href="https://www.ebi.ac.uk/research/marioni" target="_blank"> Marioni Lab </a>**

<br>

#### Contact
Please contact Shila Ghazanfar (shila.ghazanfar-at-cruk.cam.ac.uk) or John Marioni (marioni-at-ebi.ac.uk) with any web app-related queries or suggestions.
