---
title: "E8.5 seqFISH - Generate figure panels"
author: "Shila Ghazanfar"
date: "10/08/2020"
output:
       html_document:
                     toc: true
                     toc_float:
                           collapsed: false
                           smooth_scroll: false
                     code_folding: hide
                     fig_width: 16
                     fig_height: 12
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE,
                      cache = TRUE,cache.lazy = FALSE)
```

```{r}
set.seed(2020)
```

```{r}
library(SingleCellExperiment)
library(scater)
library(scran)
library(ComplexHeatmap)
library(ggplot2)
library(ggrepel)
library(GGally)
library(limma)
library(patchwork)
library(reshape)
library(gtools)
library(scattermore)
library(gganimate)
library(ggmap)
library(randomForest)
library(batchelor)
library(igraph)
library(cowplot)
library(ggpubr)
library(Matrix)
library(grDevices)
library(reshape)
library(DelayedArray)
library(HDF5Array)
library(princurve)
library(e1071)
library(openxlsx)
library(nord)
```

```{r}
sapply(list.files("../functions/", full.names = TRUE), source)
```

# fastMNN expression of each embryo and gene

```{r}
sce = readRDS("../analysis_output/E8.5/E8.5_sce_filt_unlabelled.Rds")
sce
mnn_out = readRDS("../analysis_output/E8.5/E8.5_fastMNN.Rds")
mnn_out

sce <- reNormalise(sce)

assay(sce, "fastMNN") <- assay(mnn_out, "reconstructed")
colnames(assay(sce, "fastMNN")) <- colnames(sce)

if (!file.exists("../analysis_output/E8.5/libsize_MNN.Rds")) {
  reducedDim(sce, "libsize_MNN") <- reducedDim(mnn_out, "corrected")
  reducedDim(sce, "libsize_MNN_UMAP") <- calculateUMAP(t(reducedDim(sce, "libsize_MNN")))
} else {
  saveRDS(reducedDim(sce, "libsize_MNN"), file = "../analysis_output/E8.5/libsize_MNN.Rds")
  saveRDS(reducedDim(sce, "libsize_MNN_UMAP"), file = "../analysis_output/E8.5/libsize_MNN_UMAP.Rds")
}
```

# Add cell type and clustering information

```{r}
mapping_dt = updateMesenchymeLabel(readRDS("../analysis_output/E8.5/celltype_annotation_refined.Rds"))
colData(sce) <- cbind(colData(sce), mapping_dt[colnames(sce), setdiff(colnames(mapping_dt),colnames(colData(sce)))])

celltype_cluster_dt = readRDS("../analysis_output/E8.5/libsize_clustering.Rds")
colData(sce)$cluster <- celltype_cluster_dt[rownames(celltype_cluster_dt),]$cluster
colData(sce)$cluster_sub <- celltype_cluster_dt[rownames(celltype_cluster_dt),]$cluster_sub
```

# Load mRNA data

```{r}
mRNA_df = readRDS("../analysis_output/E8.5/E8.5_mRNA_df_unfiltered.Rds")
dim(mRNA_df)
head(mRNA_df)
```

# Expression colour legend

```{r}
df = data.frame(x = 1:10)
g = ggplot(df, aes(x = x, y = x)) + 
  geom_point(aes(colour = x)) + 
  scale_colour_gradient2(low = "gray75", mid = "cornflowerblue", high = "black",
                         midpoint = 5,
                         labels = c("Low", "High"), 
                         breaks = c(1,10),
                         limits = c(1,10)) + 
  guides(colour = guide_colourbar(title = "")) + 
  theme(legend.position = "bottom")
g

gg = as_ggplot(get_legend(g))
ggsave(gg, file = "../Figures/expression_legend.pdf", height = 1, width = 2)
```

# Imputed expression colour legend

```{r}
df = data.frame(x = 1:10)
g = ggplot(df, aes(x = x, y = x)) + 
  geom_point(aes(colour = x)) + 
  scale_colour_gradient2(low = "gray75", mid = "#9B6A12", high = "black",
                         midpoint = 5,
                         labels = c("Low", "High"), 
                         breaks = c(1,10),
                         limits = c(1,10)) + 
  guides(colour = guide_colourbar(title = "")) + 
  theme(legend.position = "bottom")
g

gg = as_ggplot(get_legend(g))
ggsave(gg, file = "../Figures/expression_imputed_legend.pdf", height = 1, width = 2)
```

# Add smoothed per-celltype expression as an assay

```{r}
smoothinc = readRDS("../analysis_output/E8.5/smoothinc.Rds")
assay(sce, "smoothinc") <- smoothinc[rownames(sce), colnames(sce)]
```

# Key summary statistics

Number of genes and standard error etc

```{r}
# mean number of molecules detected over the three embryos
mean(tapply(sce$total, sce$embryo, function(x) mean(x)))
# with standard error
sd(tapply(sce$total, sce$embryo, function(x) mean(x)))/sqrt(3)

# mean number of detected genes over the three embryos
mean(tapply(sce$detected, sce$embryo, function(x) mean(x)))
# with standard error
sd(tapply(sce$detected, sce$embryo, function(x) mean(x)))/sqrt(3)

# proportion of detected mRNA belonging to cell regions
mean(mRNA_df$uniqueID %in% colnames(sce))
```

Number of genes split by germ layers

```{r}
# split genes based on their highest expression for each broad germ layer
atlas_sub = readRDS("../analysis_output/E8.5/atlas_sub.Rds")

broad_celltypes = list(
  "Extraembryonic" = c("Allantois",
                       "ExE mesoderm",
                       "PGC"),
  "Ectoderm" = c("Neural crest",
                 "Spinal cord",
                 "Surface ectoderm",
                 "Forebrain/Midbrain/Hindbrain"),
  "Endoderm" = c("Definitive endoderm",
                 "Def. endoderm",
                 "Gut",
                 "Gut tube",
                 "Parietal endoderm",
                 "ExE endoderm"),
  "Mesoderm" = c("Blood progenitors",
                 "Blood progenitors 2",
                 "Blood progenitors 1",
                 "Cardiomyocytes",
                 "Erythroid1",
                 "Erythroid2",
                 "Erythroid3",
                 "Erythroid",
                 "Haematoendothelial progenitors",
                 "Mesenchyme",
                 "Mixed mesenchymal mesoderm",
                 "NMP",
                 "Paraxial mesoderm",
                 "Pharyngeal mesoderm",
                 "Somitic mesoderm",
                 "Endothelium",
                 "Intermediate mesoderm",
                 "Lateral plate mesoderm",
                 "Splanchnic mesoderm",
                 "Cranial mesoderm",
                 "Anterior somitic tissues",
                 "Presomitic mesoderm",
                 "Dermomyotome",
                 "Caudal Mesoderm",
                 "Sclerotome"),
  "Low quality" = "Low quality"
)

broad_celltypes
broad_celltypes_long = rep(names(broad_celltypes), times = unlist(lapply(broad_celltypes,length)))
names(broad_celltypes_long) <- unlist(broad_celltypes)

# assign each atlas cell to the germ layer

atlas_sub_germ = broad_celltypes_long[atlas_sub$celltype]

counts_germ_sample = sumCountsAcrossCells(logcounts(atlas_sub),
                                          interaction(atlas_sub_germ, atlas_sub$sample),
                                          average = TRUE)
head(counts_germ_sample)

# for each sample now calculate the mean logcounts
counts_germ = sumCountsAcrossCells(counts_germ_sample,
                                   gsub("\\..*", "", colnames(counts_germ_sample)),
                                   average = TRUE,
                                   assay.type = "average")
germ_genes = colnames(counts_germ)[apply(assay(counts_germ, "average"), 1, function(x) which(x == max(x)))]
names(germ_genes) <- rownames(counts_germ)

germlayers = unique(germ_genes)

num_profiled = NULL
num_represented = NULL
prop_represented = NULL
num_represented_se = NULL

for (germ in germlayers) {
  
  sce_sub_sub = sce[names(germ_genes)[germ_genes == germ], ]
  sce_sub_sub_qc <- scater::perCellQCMetrics(sce_sub_sub)
  
  num_profiled[germ] = nrow(sce_sub_sub)
  
  # mean number of detected genes over the three embryos
  num_represented[germ] = mean(tapply(sce_sub_sub_qc$detected, sce$embryo, function(x) mean(x)))
  
  # mean proportion of the profiled genes
  prop_represented[germ] = num_represented[germ]/nrow(sce_sub_sub)
  
  # with standard error
  num_represented_se[germ] = sd(tapply(sce_sub_sub_qc$detected, sce$embryo, function(x) mean(x)))/sqrt(3)
  
}

germlayer_representation = cbind(num_profiled, num_represented, prop_represented, num_represented_se, num_represented_per_num_profiled = num_represented_se/num_profiled)
write.table(germlayer_representation, file = "../analysis_output/E8.5/germlayer_representation.tsv",
            sep = "\t", quote = FALSE, row.names = TRUE, col.names = TRUE)
```

# Plot joint UMAP with atlas

```{r}
joint_UMAP = readRDS("../analysis_output/E8.5/E8.5_joint_UMAP_corrected_libsize.Rds")
atlas_sub = readRDS("../analysis_output/E8.5/atlas_sub.Rds")
atlas_sub$celltype_parsed <- updateMesenchymeLabel(data.frame(x = atlas_sub$celltype_parsed))[,1]
joint_subclusters = readRDS("../analysis_output/E8.5/E8.5_joint_subclustering_libsize.Rds")

joint_UMAP_df = data.frame(
  cell = rownames(joint_UMAP),
  UMAP1 = joint_UMAP[,1],
  UMAP2 = joint_UMAP[,2],
  type = ifelse(substring(rownames(joint_UMAP),1,1) == "e", "seqFISH", "Atlas"),
  embryo = factor(colData(sce)[rownames(joint_UMAP),"embryo"], 
                  levels = c("embryo1", "embryo2", "embryo3", "Atlas")),
  subcluster = joint_subclusters[rownames(joint_UMAP)]
)
joint_UMAP_df[is.na(joint_UMAP_df$embryo), "embryo"] <- "Atlas"
rownames(joint_UMAP_df) <- as.character(joint_UMAP_df$cell)
joint_UMAP_df$celltype <- NA
joint_UMAP_df[colnames(sce), "celltype"] <- as.character(sce$celltype_mapped_refined)
joint_UMAP_df[colnames(atlas_sub), "celltype"] <- as.character(atlas_sub$celltype_parsed)

joint_UMAP_df$type <- factor(joint_UMAP_df$type, levels = c("Atlas", "seqFISH"),
                             labels = c("Gastrulation atlas", "seqFISH")
)

g = ggplot(subset(joint_UMAP_df, !celltype %in% "Low quality"), aes(x = UMAP1, y = UMAP2, colour = celltype)) + 
  geom_point(size = 0.5) + 
  scale_colour_manual(values = celltype_colours) + 
  theme_classic() + 
  theme(legend.position = "none") + 
  theme(axis.text = element_blank()) + 
  theme(axis.ticks = element_blank()) +
  facet_wrap(~type) +
  coord_fixed() +
  theme(strip.background = element_rect(colour = "white")) +
  theme(strip.text = element_text(size = 20)) +
  NULL
g
ggsave(g, file = "../Figures/joint_umap_celltype_split.pdf",
       height = 4, width = 7)

# extract the legend
gg = as_ggplot(get_legend(g + theme(legend.position = "right") + 
                            guides(colour = guide_legend(title = "",
                                                         ncol = 1,
                                                         override.aes = list(size = 5)))))
print(gg)

g = ggplot(joint_UMAP_df[sample(rownames(joint_UMAP_df)),], aes(x = UMAP1, y = UMAP2, colour = type)) + 
  geom_point(size = 0.5, alpha = 0.3) + 
  scale_colour_manual(values = c("Gastrulation atlas" = "orange", "seqFISH" = "navy")) + 
  theme_classic() + 
  theme(legend.position = "bottom") + 
  theme(axis.text = element_blank()) + 
  theme(axis.ticks = element_blank()) +
  coord_fixed() +
  guides(colour = guide_legend(title = "",
                               override.aes = list(size = 8, alpha = 1))) +
  NULL
g
ggsave(g, file = "../Figures/joint_umap_type.pdf",
       height = 5, width = 5)


g = ggplot(joint_UMAP_df[sample(rownames(joint_UMAP_df)),], aes(x = UMAP1, y = UMAP2, colour = embryo)) + 
  geom_point(size = 0.5, alpha = 0.3) + 
  theme_classic() + 
  theme(legend.position = "bottom") + 
  theme(axis.text = element_blank()) + 
  theme(axis.ticks = element_blank()) +
  coord_fixed() +
  guides(colour = guide_legend(title = "",
                               override.aes = list(size = 8, alpha = 1))) +
  NULL
g
ggsave(g, file = "../Figures/joint_umap_embryo.pdf",
       height = 5, width = 5)

gg = g + facet_wrap(~embryo)
gg
ggsave(gg, file = "../Figures/joint_umap_embryo_split.pdf",
       height = 6, width = 5)

joint_UMAP_df_subcluster = data.frame(
  subcluster = sort(unique(joint_UMAP_df$subcluster))
)
joint_UMAP_df_subcluster$UMAP1 <- tapply(joint_UMAP_df$UMAP1, joint_UMAP_df$subcluster, mean)[joint_UMAP_df_subcluster$subcluster]
joint_UMAP_df_subcluster$UMAP2 <- tapply(joint_UMAP_df$UMAP2, joint_UMAP_df$subcluster, mean)[joint_UMAP_df_subcluster$subcluster]

g = ggplot(joint_UMAP_df[sample(rownames(joint_UMAP_df)),], aes(x = UMAP1, y = UMAP2, colour = subcluster)) + 
  geom_point(size = 0.5, alpha = 0.3) + 
  theme_classic() + 
  theme(legend.position = "none") + 
  theme(axis.text = element_blank()) + 
  theme(axis.ticks = element_blank()) +
  scale_colour_manual(values = sample(rainbow(length(unique(joint_UMAP_df$subcluster)))),
                      aesthetics = c("fill", "colour")) +
  coord_fixed() +
  guides(colour = guide_legend(title = "",
                               override.aes = list(size = 8, alpha = 1))) +
  geom_label_repel(colour = "black", data = joint_UMAP_df_subcluster, aes(label = subcluster, fill = subcluster), alpha = 0.6) +
  NULL
g
ggsave(g, file = "../Figures/joint_umap_joint_subclusters.pdf",
       height = 10, width = 10)

# violin plot of joint mapping scores for the integration
qual_order = names(sort(tapply(sce$celltype_mapping_score, sce$celltype_mapped_refined, mean), decreasing = TRUE))
df = as.data.frame(colData(sce))
df <- df[df$celltype_mapped_refined != "Low quality",]
df$celltype_mapped_refined <- factor(df$celltype_mapped_refined, levels = qual_order)
numcells_df = data.frame(
  celltype_mapped_refined = sort(unique(df$celltype_mapped_refined)),
  numCells = as.numeric(table(df$celltype_mapped_refined)[sort(unique(df$celltype_mapped_refined))])
)
g = ggplot(df, 
           aes(x = celltype_mapped_refined, y = celltype_mapping_score)) + 
  theme_classic() + 
  geom_violin(aes(fill = celltype_mapped_refined),
              scale = "width", draw_quantiles = 0.5) + 
  scale_fill_manual(values = celltype_colours) + 
  theme(legend.position = "none") + 
  xlab("") +
  ylab("Mapping score") + 
  ylim(c(0,1.3)) +
  theme(axis.text.x = element_text(angle = 60, hjust = 1)) + 
  geom_text(aes(x = celltype_mapped_refined, label = numCells),
            y = 1.05, data = numcells_df,
            inherit.aes = FALSE, angle = 30, hjust = 0) + 
  theme(plot.margin = unit(c(2,2,0,1), "mm")) +
  NULL
g
ggsave(g, file = "../Figures/mapping_score_atlas.pdf", height = 4, width = 8)
```

```{r}
boundary_polygons = getSegmentationVerticesDF(colData(sce),
                                              xname = "segmentation_vertices_x_global_affine",
                                              yname = "segmentation_vertices_y_global_affine",
                                              othercols = c("uniqueID","z","cluster","cluster_sub","embryo", "celltype_mapped_refined"))
```

# Define plotEmbryo function

Currently, function requires boundary_polygons to be defined in global 
environment.

```{r}
plotEmbryo = function(
  sce,
  assayName = "fastMNN",
  embryo_subset = c("embryo1"),
  z_subset = c(2),
  colours = c("gray75", "cornflowerblue", "black"),
  gene_name = "Ttn",
  quantileNorm = FALSE,
  outline = FALSE
) {
  
  title.size = 15
  
  boundary_polygons_bkg = subset(
    boundary_polygons,
    embryo %in% embryo_subset & 
      z %in% z_subset
  )
  
  boundary_polygons_sub = boundary_polygons_bkg
  
  selectedNames = boundary_polygons_bkg$uniqueID
  
  pc_vals = assay(sce, assayName)[gene_name, ]
  names(pc_vals) <- colnames(sce)
  
  pc_vals <- pc_vals[sce$uniqueID %in% selectedNames]
  
  if (quantileNorm) {
    # quantile normalise to the logcounts
    pc_vals_pre = pc_vals
    logcounts_vals = assay(sce, "logcounts")[gene_name, names(pc_vals)]
    pc_vals[order(pc_vals)] <- sort(logcounts_vals)
    pc_vals[logcounts_vals == 0] <- 0
    
    
    if (FALSE) {
      par(mfrow=c(1,4))
      
      plot(logcounts_vals,
           pc_vals_pre, 
           ylab = "Reconstructed MNN",
           xlab = "logcounts")
      abline(c(0,1), col = "red")
      
      plot(logcounts_vals,
           pc_vals,
           xlab = "logcounts",
           ylab = "Quantile normalised reconstructed MNN")
      abline(c(0,1), col = "red")
      
      plot(sort(logcounts_vals),
           sort(pc_vals),
           xlab = "sorted logcounts",
           ylab = "sorted Quantile normalised reconstructed MNN")
      abline(c(0,1), col = "red")
      
      plot(pc_vals_pre,
           pc_vals, 
           xlab = "Reconstructed MNN",
           ylab = "Quantile normalised reconstructed MNN")
      abline(c(0,1), col = "red")
      
    }
  }
  
  pc_vals[pc_vals < 0] <- 0
  
  if (length(pc_vals) == 0) stop("No cells are selected for subsetting, please tick at least one option for each of the embryo, z-stack and mapped cell type categories.")
  if (all(pc_vals == 0)) {
    pc_cols = rep(colours[1], length(pc_vals))
  } else {
    pc_cols = colorRampPalette(colours)(100)[as.numeric(cut(pc_vals,100))]
  }
  names(pc_cols) <- names(pc_vals)
  
  g_base = ggplot()
  
  if (!is.logical(outline)) {
    colval = outline
  } else {
    colval = ifelse(outline, "black", NA)
  }
  
  g = g_base +
    geom_polygon(aes(x = segmentation_vertices_x_global_affine,
                     y = -segmentation_vertices_y_global_affine,
                     group = uniqueID,
                     fill = uniqueID),
                 # colour = "black", size = 0.1,
                 colour = colval, size = 0.1,
                 data = boundary_polygons_sub) +
    facet_grid(z~embryo) +
    scale_fill_manual(values = pc_cols) +
    theme_classic() +
    coord_fixed() +
    theme(legend.position = "none") + 
    ggtitle("") +
    xlab("") +
    theme(axis.line = element_blank()) + 
    theme(axis.text = element_blank()) + 
    theme(axis.title.y = element_blank()) +
    theme(axis.ticks = element_blank()) +
    theme(strip.background = element_blank()) +
    theme(strip.text = element_blank()) +
    theme(plot.title = element_text(hjust = 0.5, size = title.size)) +
    NULL
  
  g_spatial <- g
  
  return(g_spatial)
  
}
```

Plot for various genes


```{r}
sapply(rownames(sce), function(gene){
  print(gene)
  g = plotEmbryo(sce, 
                 gene_name = gene, 
                 assayName = "fastMNN",
                 quantileNorm = TRUE,
                 embryo_subset = "embryo1",
                 z_subset = "2") +
    NULL
  if (gene == "Ttn") {
    g <- g + add_scalebar(border.size = 10)
  }
  ggsave(g, file = paste0("../Figures/spatial_plot/corrected_", gene, "_embryo1_z2.pdf"),
         height = 20, width = 24)
})

genes = c("Ttn", "Hoxa9", "Ezh2")
gList = sapply(genes, function(gene) {
  g = plotEmbryo(sce, gene_name = gene, assayName = "logcounts") + 
    ggtitle(paste0("logcounts, ", gene)) +
    plotEmbryo(sce, gene_name = gene, assayName = "fastMNN",
               quantileNorm = FALSE) + ggtitle("fastMNN") + 
    plotEmbryo(sce, gene_name = gene, assayName = "fastMNN",
               quantileNorm = TRUE) + ggtitle("fastMNN with quantile") +
    plot_layout(nrow = 1)
  return(g)
}, simplify = FALSE)

wrap_plots(gList, ncol = 1)


# plot smoothed expression of hox genes
genes = mixedsort(grep("Hox", rownames(sce), value = "TRUE"))
genesList = split(genes, substring(genes, 4,4))
gList = sapply(genesList, function(genes) {
  sapply(genes, function(gene) {
    print(gene)
    gAll = sapply(unique(sce$embryo), function(emb){
      g = plotEmbryo(sce, 
                     gene_name = gene, 
                     assayName = "smoothinc",
                     quantileNorm = TRUE,
                     embryo_subset = emb,
                     z_subset = "2") +
        theme(plot.title = element_text(face = "italic")) +
        NULL
      if (gene == genes[1]) {
        colval = "black"
      } else {
        colval = "white"
      }
      g <- g + add_scalebar(box.col = colval)
      if (emb == "embryo1") {
        g <- g + ggtitle(gene) 
      }
      return(g)
    }, simplify = FALSE)
    g <- wrap_plots(gAll, ncol = 1)
    return(g)
  }, simplify = FALSE)
}, simplify = FALSE)

lapply(gList, length)

# test
wrap_plots(gList[[1]], nrow = 1)

# all
gList1 = lapply(gList, function(gList_1) wrap_plots(gList_1, nrow = 1))

sapply(names(gList), function(gL){
  print(gL)
  g = wrap_plots(gList[[gL]], nrow = 1)
  ggsave(g, file = paste0("../Figures/Hox",gL, "_smoothed.png"),
         height = 10, width = 2*length(gList[[gL]]))
})

# and just regular for comparison
genes = mixedsort(grep("Hox", rownames(sce), value = "TRUE"))
genesList = split(genes, substring(genes, 4,4))
gList = sapply(genesList, function(genes) {
  sapply(genes, function(gene) {
    print(gene)
    gAll = sapply(unique(sce$embryo), function(emb){
      g = plotEmbryo(sce, 
                     gene_name = gene, 
                     assayName = "fastMNN",
                     quantileNorm = TRUE,
                     embryo_subset = emb,
                     z_subset = "2") +
        theme(plot.title = element_text(face = "italic")) +
        NULL
      if (gene == genes[1]) {
        colval = "black"
      } else {
        colval = "white"
      }
      g <- g + add_scalebar(box.col = colval)
      if (emb == "embryo1") {
        g <- g + ggtitle(gene) 
      }
      return(g)
    }, simplify = FALSE)
    g <- wrap_plots(gAll, ncol = 1)
    return(g)
  }, simplify = FALSE)
}, simplify = FALSE)

sapply(names(gList), function(gL){
  print(gL)
  g = wrap_plots(gList[[gL]], nrow = 1)
  ggsave(g, file = paste0("../Figures/Hox",gL, ".png"),
         height = 10, width = 2*length(gList[[gL]]))
})
```


```{r}
# and for imputed gene expression
file_combined = "../analysis_output/imputation/combined.h5"
cnames = readRDS("../analysis_output/imputation/combined_cnames.Rds")
rnames = readRDS("../analysis_output/imputation/combined_rnames.Rds")

imp <- HDF5Array(filepath = file_combined, name = "logcounts")
rownames(imp) <- rnames
colnames(imp) <- cnames

genes_imp = rnames



genes = mixedsort(grep("os|aas", grep("Hox", rnames, value = TRUE), value = TRUE, invert = TRUE))

sce_hox_imp = SingleCellExperiment(assay = list(imputed = imp[genes,colnames(sce)]))
colData(sce_hox_imp) <- colData(sce)

genesList = split(genes, substring(genes, 4,4))
gList = sapply(genesList, function(genes) {
  sapply(genes, function(gene) {
    print(gene)
    gAll = sapply(unique(sce_hox_imp$embryo), function(emb){
      g = plotEmbryo(sce_hox_imp, 
                     gene_name = gene, 
                     assayName = "imputed",
                     quantileNorm = FALSE,
                     embryo_subset = emb,
                     z_subset = "2") +
        theme(plot.title = element_text(face = "italic")) +
        NULL
      if (gene == genes[1]) {
        colval = "black"
      } else {
        colval = "white"
      }
      g <- g + add_scalebar(box.col = colval)
      if (emb == "embryo1") {
        g <- g + ggtitle(gene) +
          theme(plot.title = element_text(colour = ifelse(gene %in% rownames(sce), "black", "red")))
      }
      return(g)
    }, simplify = FALSE)
    g <- wrap_plots(gAll, ncol = 1)
    return(g)
  }, simplify = FALSE)
}, simplify = FALSE)

lapply(gList, length)

sapply(names(gList), function(gL){
  print(gL)
  g = wrap_plots(gList[[gL]], nrow = 1)
  ggsave(g, file = paste0("../Figures/Hox",gL, "_imputed.png"),
         height = 10, width = 2*length(gList[[gL]]))
})
```

Plot for the other embryos

```{r}
genesToPlot = c("Ttn", "Popdc2", "Hand1", "Gata5", "Six3", "Lhx2", "Otx2", "Pou3f1", "Sox2", "Foxf1", "Foxa1", "Cldn4")
gList = list()

for (embryoval in c("embryo1", "embryo2", "embryo3")) {
  for (gene in genesToPlot) {
    g = plotEmbryo(sce, 
                   gene_name = gene,
                   assayName = "fastMNN",
                   quantileNorm = TRUE,
                   embryo_subset = embryoval,
                   z_subset = "2") + 
      ggtitle(gene) + 
      theme(plot.title = element_text(face = "italic")) +
      theme(plot.margin = unit(c(0,0,0,0), "cm")) + 
      NULL
    if (gene == genesToPlot[1]) { 
      colval = "black"
    } else {
      colval = "white"
    }
    g = g + add_scalebar(box.col = colval)
    gList[[embryoval]][[gene]] <- g 
  }
}

ggsave(wrap_plots(gList[["embryo1"]], nrow = 3), file = "../Figures/spatial_expression_embryo1.pdf", height = 10, width = 8)

ggsave(wrap_plots(gList[["embryo2"]], nrow = 3), file = "../Figures/spatial_expression_embryo2.pdf", height = 10, width = 8)

ggsave(wrap_plots(gList[["embryo3"]], nrow = 3), file = "../Figures/spatial_expression_embryo3.pdf", height = 10, width = 8)
```

# Cell type graph

```{r}
celltype_mapped_col <- celltype_colours[as.character(sce$celltype_mapped_refined)]
names(celltype_mapped_col) <- colnames(sce)

celltype_mapped_conf = sce$celltype_mapping_score
names(celltype_mapped_conf) <- colnames(sce)

embryovals = unique(boundary_polygons$embryo)
zvals = unique(boundary_polygons$z)

for (embryoval in embryovals) {
  for (zval in zvals) {
    
    g = ggplot(subset(boundary_polygons, embryo == embryoval & z == zval)) + 
      geom_polygon(aes(group = uniqueID,
                       x = segmentation_vertices_x_global_affine,
                       y = -segmentation_vertices_y_global_affine,
                       fill = uniqueID), colour = "black",
                   size = 0.1) + 
      scale_fill_manual(values = celltype_mapped_col) +
      theme_classic() +
      coord_fixed() +
      theme(legend.position = "none") +
      NULL
    g
    
    gg = g + theme(axis.line = element_blank()) + 
      theme(axis.text = element_blank()) +
      theme(axis.ticks = element_blank()) + 
      theme(axis.title = element_blank()) +
      add_scalebar() +
      NULL
    gg
    ggsave(gg, file = paste0("../Figures/celltype_mapped_refined_",
                             embryoval, "_", zval, ".png"),
           height = 10, width = 8 )
  }
}
```

# Cell Cell Contact Map construction

```{r}
graph = readRDS("../analysis_output/E8.5/E8.5_neighbourGraph_1.3.Rds")

nperm = 500

out_all = cellCellContact(sce, group = "celltype_mapped_refined", graph, nperm = nperm)
g = cellCellContactMap(out_all, exclude = c("Low quality", "ExE endoderm"))
g

split_categories = interaction(sce$embryo, sce$z)
sce_split = splitSCE(sce, split_categories)

outList = lapply(sce_split, cellCellContact, group = "celltype_mapped_refined",
                 graph = graph, nperm = nperm, plot = FALSE)
gList = lapply(outList, cellCellContactMap, order = levels(cellCellContactMap(out_all)$data$subcluster_1),
               exclude = c("Low quality", "ExE endoderm"))

gList2 = sapply(names(gList), function(n) gList[[n]] + ggtitle(n), simplify = FALSE)
wrap_plots(gList2)

gList[[1]]

pmat_all = lapply(outList, "[[", 2)
all_celltypes = sort(unique(unlist(lapply(pmat_all,rownames))))
pmatArray = abind::abind(lapply(pmat_all, function(x) {
  x_all = matrix(NA, nrow = length(all_celltypes),
                 ncol = length(all_celltypes), dimnames = list(all_celltypes, all_celltypes))
  x_all[rownames(x), colnames(x)] <- x
  return(x_all)
}), along = 3)
pmatMean = apply(pmatArray, 1:2, mean, na.rm = TRUE)
g = cellCellContactMap(list(obs = pmatMean, pmat = pmatMean),
                       exclude = c("Low quality", "ExE endoderm"))
g

hc = hclust(dist(1*(pmatMean > 0.95)))
g = cellCellContactMap(list(obs = pmatMean, pmat = pmatMean),
                       exclude = c("Low quality", "ExE endoderm"),
                       order = hc$labels[hc$order])
g

h = cellCellContactHeatmap(list(obs = pmatMean, pmat = pmatMean),
                           col = celltype_colours,
                           # exclude = c("Low quality"),
                           order = hc$labels[hc$order],
                           factor = 0.5)
h

h = cellCellContactHeatmap(list(obs = pmatMean, pmat = pmatMean),
                           col = celltype_colours,
                           exclude = c("Low quality", names(which(table(sce$celltype_mapped_refined) < 100))),
                           factor = 0.5,
                           split_n = 12)
h

h = cellCellContactHeatmapTriangle(list(obs = pmatMean, pmat = pmatMean),
                                   col_ann = celltype_colours,
                                   exclude = c("Low quality", names(which(table(sce$celltype_mapped_refined) < 100))),
                                   factor = 0.5,
                                   split_n = 12)
h

pdf("../Figures/neighbourhood_celltype_mapped_refined.pdf",
    height = 8, width = 8)
print(h)
dev.off()

saveRDS(pmatArray, file = "../analysis_output/E8.5/pmatArray.Rds")
```

# cell cell separability

Separability is calculated by fitting a SVM model to discriminate between
the cell types in physical space, and returning the resubstitution 
classification error rate. More errors means more mixing of the cell types.

```{r}
dist_sep = cellCellSeparability(sce, split = interaction(sce$embryo, sce$z),
                                group = "celltype_mapped_refined",
                                coordNames = c("x_global_affine", 
                                               "y_global_affine"),
                                plot = FALSE)

g = cellCellSeparabilityMap(dist_sep, exclude = "Low quality")
ggsave(g, file = "../Figures/separability_celltype_mapped_refined.pdf",
       height = 8, width = 8)
```

# Use the PAGA graph to highlight expression of the atlas genes as markers

```{r}
dist_raw = read.csv("../ricard/PAGA_distances.csv", header = TRUE, row.names = 1)
rownames(dist_raw)[rownames(dist_raw) == "Forebrain Midbrain Hindbrain"] <- "Forebrain/Midbrain/Hindbrain"
dist = as.matrix(dist_raw)
colnames(dist) <- rownames(dist)

hc = hclust(as.dist(dist), method = "ward.D")
plot(hc)

atlas = readRDS("../analysis_output/E8.5/atlas_for_celltype_mapping.Rds")

genes_raw = c("Pitx1 Hoxa9 Hand1 Gata5 Foxf1 Otx2 Pou3f1 Sox2 Ttn Popdc2 Cldn4 Lhx2 Foxa1 Fezf1 Six3")
genes = unlist(strsplit(genes_raw, " "))
genes

propExpressed = apply(logcounts(atlas)[genes,],1,function(x){
  tapply(x, atlas$celltype, function(x) mean(x))
})
propExpressed <- apply(propExpressed, 2, function(x) x/max(x))

hc = hclust(as.dist(dist[rownames(propExpressed), rownames(propExpressed)]), method = "ward.D")
plot(hc)

h = Heatmap(t(propExpressed), cluster_columns = hc,
            col = c("gray75", "cornflowerblue", "black"),
            column_dend_height = unit(30, "mm"),
            column_split = 10,
            top_annotation = HeatmapAnnotation(
              which = "column",
              celltype = names(celltype_colours[rownames(propExpressed)]),
              col = list(celltype = celltype_colours[rownames(propExpressed)]),
              annotation_legend_param = list(celltype = list(title = "Cell type"))
            ),
            heatmap_legend_param = list(title = "Relative\nmean\nexpression",
                                        hjust = 0.5),
            column_title = "Relative mean expression of selected genes in Pijuan-Sala et al (2019)\nMouse Gastrulation Atlas dataset (E8.5 cells)"
)

h

pdf("../Figures/atlas_heatmap_selectedgenes.pdf",
    height = 10, width = 15)
print(h)
dev.off()
```

Same for the smFISH genes

```{r}
genes = rownames(read.delim("../Data/smFISH/channel_info/smFISH__gene_key__new.tab",
                            header = TRUE, row.names = 1))
genes %in% rownames(atlas)
# get expression dotplot for these genes and celltypes

propExpressed = apply(logcounts(atlas)[genes,],1,function(x){
  tapply(x, atlas$celltype, function(x) mean(x))
})
propExpressed <- apply(propExpressed, 2, function(x) x/max(x))

hc = hclust(as.dist(dist[rownames(propExpressed), rownames(propExpressed)]), method = "ward.D")
plot(hc)

h = Heatmap(t(propExpressed), cluster_columns = hc,
            col = c("gray75", "cornflowerblue", "black"),
            column_dend_height = unit(30, "mm"),
            column_split = 10,
            top_annotation = HeatmapAnnotation(
              which = "column",
              celltype = names(celltype_colours[rownames(propExpressed)]),
              col = list(celltype = celltype_colours[rownames(propExpressed)]),
              annotation_legend_param = list(celltype = list(title = "Cell type"))
            ),
            heatmap_legend_param = list(title = "Relative\nmean\nexpression",
                                        hjust = 0.5),
            column_title = "Relative mean expression of genes in Pijuan-Sala et al (2019)\nMouse Gastrulation Atlas dataset (E8.5 cells)"
)

h

pdf("../Figures/atlas_heatmap_smFISHgenes.pdf",
    height = 10, width = 15)
print(h)
dev.off()
```

# Graph embryo with celltype colour split into endo ecto mesoderm germ layers

```{r}
broad_celltypes = list(
  "Extraembryonic" = c("Allantois",
                       "ExE mesoderm",
                       "PGC"),
  "Ectoderm" = c("Neural crest",
                 "Spinal cord",
                 "Surface ectoderm",
                 "Forebrain/Midbrain/Hindbrain"),
  "Endoderm" = c("Definitive endoderm",
                 "Def. endoderm",
                 "Gut",
                 "Gut tube",
                 "Parietal endoderm",
                 "ExE endoderm"),
  "Mesoderm" = c("Blood progenitors",
                 "Blood progenitors 2",
                 "Blood progenitors 1",
                 "Cardiomyocytes",
                 "Erythroid1",
                 "Erythroid2",
                 "Erythroid3",
                 "Erythroid",
                 "Haematoendothelial progenitors",
                 "Mesenchyme",
                 "NMP",
                 "Paraxial mesoderm",
                 "Pharyngeal mesoderm",
                 "Somitic mesoderm",
                 "Endothelium",
                 "Intermediate mesoderm",
                 "Lateral plate mesoderm",
                 "Splanchnic mesoderm",
                 "Cranial mesoderm",
                 "Anterior somitic tissues",
                 "Presomitic mesoderm",
                 "Dermomyotome",
                 "Caudal Mesoderm",
                 "Sclerotome"),
  "Low quality" = "Low quality"
)

broad_celltypes_labels = rep(names(broad_celltypes), times = unlist(lapply(broad_celltypes, length)))
names(broad_celltypes_labels) <- unlist(broad_celltypes)

broad_celltypes_colours = c(RColorBrewer::brewer.pal(6, "Set3")[c(2,6,4,5)], "#ffffff")
names(broad_celltypes_colours) <- names(broad_celltypes)

broad_celltype_colours_values = broad_celltypes_colours[rep(names(broad_celltypes), times = unlist(lapply(broad_celltypes, length)))]
names(broad_celltype_colours_values) <- unlist(broad_celltypes)

boundary_polygons = getSegmentationVerticesDF(colData(sce),
                                              xname = "segmentation_vertices_x_global_affine",
                                              yname = "segmentation_vertices_y_global_affine",
                                              othercols = c("uniqueID","z","celltype_mapped_refined","cluster","cluster_sub","embryo"))

setdiff(boundary_polygons$celltype_mapped_refined, names(broad_celltypes_labels))

g = ggplot(boundary_polygons,
           aes(x = segmentation_vertices_x_global_affine,
               y = -segmentation_vertices_y_global_affine,
               fill = celltype_mapped_refined)) + 
  geom_polygon(colour = NA, size = 0.01, aes(group = uniqueID)) + 
  theme_classic() +
  scale_fill_manual(values = broad_celltype_colours_values, na.value = "grey") +
  theme(legend.position = "none") +
  facet_grid(z~embryo) +
  theme(axis.line = element_blank()) + 
  theme(axis.ticks = element_blank()) + 
  theme(axis.text = element_blank()) +
  coord_fixed() +
  xlab("") +
  ylab("") +
  add_scalebar() +
  NULL

print(g)

ggsave(g, file = paste0("../Figures/celltype_dermlayers.pdf"), height = 20, width = 24)

broad_celltypes_representative = unlist(lapply(broad_celltypes, "[", 1))

boundary_polygons_sub = boundary_polygons[(!duplicated(boundary_polygons$celltype_mapped_refined)) &
                                            boundary_polygons$celltype_mapped_refined %in% broad_celltypes_representative,]
boundary_polygons_sub$celltype_mapped_refined <- factor(boundary_polygons_sub$celltype_mapped_refined,
                                                        levels = broad_celltypes_representative[c("Endoderm",
                                                                                                  "Mesoderm",
                                                                                                  "Ectoderm",
                                                                                                  "Extraembryonic",
                                                                                                  "Low quality")])

g = ggplot(boundary_polygons_sub,
           aes(x = segmentation_vertices_x_global_affine,
               y = -segmentation_vertices_y_global_affine,
               group = uniqueID,
               fill = celltype_mapped_refined)) + 
  geom_polygon(colour = "gray75", size = 0.01) + 
  theme_classic() +
  scale_fill_manual(values = broad_celltype_colours_values, na.value = "grey",
                    labels = broad_celltypes_labels[names(broad_celltype_colours_values)], drop = TRUE) +
  theme(legend.position = "bottom") +
  guides(fill = guide_legend(title = "")) +
  theme(axis.line = element_blank()) + 
  theme(axis.ticks = element_blank()) + 
  theme(axis.text = element_blank()) +
  coord_fixed() +
  xlab("") +
  ylab("") +
  NULL

g

gg = as_ggplot(get_legend(g + theme(legend.position = "right") + 
                            guides(fill = guide_legend(title = ""))))
print(gg)

ggsave(gg, file = "../Figures/celltype_dermlayers_legend.pdf",
       height = 2, width = 2)


# plot the dermlayers with the actual celltype colour
gList = list()

embryovals = c("embryo1", "embryo2", "embryo3")
zvals = c(2,5)
dermlayers = c("Ectoderm", "Mesoderm", "Endoderm")

for (embryoval in embryovals) {
  for (zval in zvals) {
    for (dermlayer in dermlayers) {
      
      print(embryoval)
      print(zval)
      print(dermlayer)
      
      broad_celltype_colours_values_mask = celltype_colours
      broad_celltype_colours_values_mask[names(broad_celltypes_labels)[broad_celltypes_labels != dermlayer]] <- "lightgrey"
      
      g = ggplot(subset(boundary_polygons, embryo == embryoval & z == zval),
                 aes(x = segmentation_vertices_x_global_affine,
                     y = -segmentation_vertices_y_global_affine,
                     fill = celltype_mapped_refined)) + 
        geom_polygon(colour = NA, size = 0.005, aes(group = uniqueID)) + 
        theme_classic() +
        scale_fill_manual(values = broad_celltype_colours_values_mask, na.value = "grey") +
        theme(legend.position = "none") +
        ggtitle(dermlayer) +
        theme(axis.line = element_blank()) + 
        theme(axis.ticks = element_blank()) + 
        theme(axis.text = element_blank()) +
        coord_fixed() +
        xlab("") +
        ylab("") +
        theme(plot.title = element_text(hjust = 0.5, size = 40)) +
        NULL
      
      if (dermlayer == "Ectoderm") {
        g = g + add_scalebar(border.size = 3)
      }
      
      gList[[paste0(embryoval, ".", zval)]][[dermlayer]] <- g
      
    }
  }
}

sapply(names(gList), function(n) {
  g = wrap_plots(gList[[n]]) + plot_layout(nrow = 1)
  ggsave(g, file = paste0("../Figures/celltype_dermlayers_split_", n, ".pdf"), height = 10, width = 24)
})


# generate split legend
labels_split = split(names(broad_celltypes_labels), broad_celltypes_labels)
labels_split <- lapply(labels_split, function(x)
  intersect(x, names(which(table(sce$celltype_mapped_refined) >= 100))))

gList = lapply(labels_split, function(labels) {
  
  g = ggplot(aes(x = x, y = x), data = data.frame(x = labels)) + 
    geom_point(aes(colour = x)) + 
    theme_classic() +
    scale_colour_manual(values = celltype_colours) + 
    guides(colour = guide_legend(title = "", override.aes = list(size = 10))) +
    NULL
  
  gg = as_ggplot(get_legend(g))
  gg
  
  return(gg)
  
})

sapply(names(gList), function(n){
  if (n == "Low quality") return(NULL)
  ggsave(gList[[n]], file = paste0("../Figures/celltype_dermlayers_",
                                   n, "_legend.pdf"),
         height = 8, width = 3)
})

labels =  setdiff(names(which(table(sce$celltype_mapped_refined) >= 100)), "Low quality")
g = ggplot(aes(x = x, y = x), data = data.frame(x = labels)) + 
  geom_point(aes(colour = x)) + 
  theme_classic() +
  scale_colour_manual(values = celltype_colours) + 
  guides(colour = guide_legend(title = "", ncol = 1, override.aes = list(size = 7))) +
  NULL

gg = as_ggplot(get_legend(g))
gg
ggsave(gg, file = paste0("../Figures/celltype_legend.pdf"),
       height = 6.5, width = 2.5)



#####################
# generate a graph with just two celltypes highlighted
# plot the dermlayers with the actual celltype colour
gList = list()

celltypes = sort(setdiff(unique(sce$celltype_mapped_refined), "Low quality"))

boundary_polygons <- updateMesenchymeLabel(boundary_polygons)

for (embryoval in embryovals) {
  for (zval in zvals) {
    for (celltype in celltypes) {
      
      print(embryoval)
      print(zval)
      print(celltype)
      
      broad_celltype_colours_values_mask = celltype_colours
      broad_celltype_colours_values_mask[names(celltype_colours) != celltype] <- "lightgrey"
      broad_celltype_colours_values_mask[names(celltype_colours) == celltype] <- "black"
      
      g = ggplot(subset(boundary_polygons, embryo == embryoval & z == zval),
                 aes(x = segmentation_vertices_x_global_affine,
                     y = -segmentation_vertices_y_global_affine,
                     group = uniqueID,
                     fill = celltype_mapped_refined)) + 
        geom_polygon(colour = NA, size = 0.005) + 
        theme_classic() +
        scale_fill_manual(values = broad_celltype_colours_values_mask, na.value = "grey") +
        theme(legend.position = "none") +
        ggtitle(celltype) +
        theme(axis.line = element_blank()) + 
        theme(axis.ticks = element_blank()) + 
        theme(axis.text = element_blank()) +
        coord_fixed() +
        xlab("") +
        ylab("") +
        theme(plot.title = element_text(hjust = 0.5)) +
        NULL
      
      gList[[paste0(embryoval, ".", zval)]][[celltype]] <- g
      
    }
  }
}

sapply(names(gList), function(n) {
  g = wrap_plots(gList[[n]]) + plot_layout(ncol = 4)
  ggsave(g, file = paste0("../Figures/celltype_individual_split_", n, ".pdf"), height = 24, width = 18)
})
```

# Subcellular localisation of Ttn

```{r}
range_x = c(-1.9, 0.5)
range_y = c(-1.5, 0.95)

range_x_zoom = c(-1.4, -0.4)
range_y_zoom = c(-0.5, 0.5)

boundary_polygons = getSegmentationVerticesDF(colData(sce),
                                              xname = "segmentation_vertices_x_global_affine",
                                              yname = "segmentation_vertices_y_global_affine",
                                              othercols = c("uniqueID","z","celltype_mapped_refined","cluster","cluster_sub","embryo"))

colour_scheme = c("gray75","cornflowerblue","black")

mRNA_Ttn = mRNA_df[mRNA_df$z == 2 & mRNA_df$embryo == "embryo1",]

boundary_polygons_sub = subset(boundary_polygons, uniqueID %in% mRNA_Ttn$uniqueID)

boundary_polygons_sub$Ttn <- logcounts(sce)["Ttn", boundary_polygons_sub$uniqueID]

boundary_polygons_sub$Ttn_norm <- unsplit(tapply(boundary_polygons_sub$Ttn,
                                                 boundary_polygons_sub$embryo,
                                                 function(x) x/max(x)),
                                          boundary_polygons_sub$embryo)

g = ggplot(subset(boundary_polygons_sub, embryo == "embryo1" & z == 2),
           aes(x = segmentation_vertices_x_global_affine,
               y = -segmentation_vertices_y_global_affine,
               group = uniqueID,
               fill = Ttn_norm)) + 
  geom_polygon() + 
  theme_classic() +
  scale_fill_gradient2(low = colour_scheme[1], mid = colour_scheme[2], high = colour_scheme[3]) +
  guides(fill = guide_legend(title = "Ttn normalised logcounts")) +
  coord_fixed() +
  xlab("") +
  ylab("") +
  xlim(range_x) + 
  ylim(range_y) +
  theme(legend.position = "bottom") +
  NULL

g

g_dots = g + geom_point(aes(x = x_global_affine, y = -y_global_affine), 
                        data = subset(mRNA_Ttn, embryo == "embryo1" & z == 2),
                        size = 0.5,
                        colour = "red",
                        inherit.aes = FALSE)
g_dots

g_dots_box = g_dots + geom_polygon(
  aes(x = segmentation_vertices_x_global_affine,
      y = -segmentation_vertices_y_global_affine),
  data = data.frame(
    segmentation_vertices_x_global_affine = c(range_x_zoom[c(1,1,2,2)]), 
    segmentation_vertices_y_global_affine = -c(range_y_zoom[c(1,2,2,1)])),
  inherit.aes = FALSE, fill = NA, colour = "black", size = 2)
g_dots_box
ggsave(g_dots_box, file = "../Figures/Ttn_dots_box.pdf", height = 8, width = 9)

g_dots_zoom = g_dots + xlim(range_x_zoom) + ylim(range_y_zoom)
g_dots_zoom
ggsave(g_dots_zoom, file = "../Figures/Ttn_dots_zoom.pdf", height = 8, width = 9)


# with black background
g = ggplot(subset(boundary_polygons_sub, embryo == "embryo1" & z == 2),
           aes(x = segmentation_vertices_x_global_affine,
               y = -segmentation_vertices_y_global_affine,
               fill = Ttn_norm)) + 
  geom_polygon(aes(group = uniqueID), fill = "black", colour = "darkgrey", size = 0.1) +
  theme_classic() +
  scale_fill_gradient2(low = colour_scheme[1], mid = colour_scheme[2], high = colour_scheme[3]) +
  guides(fill = guide_legend(title = "Ttn normalised logcounts")) +
  xlab("") +
  ylab("") +
  xlim(range_x) +
  scale_y_continuous(expand = range_y, limits = range_y) +
  theme(panel.background = element_rect(fill = "black")) +
  theme(axis.text = element_blank()) + 
  theme(axis.ticks = element_blank()) +
  coord_fixed() +
  NULL

g

g_dots = g + 
  geom_point(aes(x = x_global_affine, y = -y_global_affine), 
             data = subset(mRNA_Ttn, embryo == "embryo1" & z == 2 & geneID == "Ttn"),
             size = 0.5,
             alpha = 0.7,
             colour = "red",
             shape = 16,
             inherit.aes = FALSE) + 
  geom_point(aes(x = x_global_affine, y = -y_global_affine), 
             data = subset(mRNA_Ttn, embryo == "embryo1" & z == 2 & geneID == "Tbx5"),
             size = 0.5,
             alpha = 0.7,
             colour = "cyan",
             shape = 16,
             inherit.aes = FALSE) + 
  geom_point(aes(x = x_global_affine, y = -y_global_affine), 
             data = subset(mRNA_Ttn, embryo == "embryo1" & z == 2 & geneID == "Cdh5"),
             size = 0.5,
             alpha = 0.7,
             colour = "green",
             shape = 16,
             inherit.aes = FALSE) + 
  geom_point(aes(x = x_global_affine, y = -y_global_affine), 
             data = subset(mRNA_Ttn, embryo == "embryo1" & z == 2 & geneID == "Dlk1"),
             size = 0.5,
             alpha = 0.7,
             colour = "orange",
             shape = 16,
             inherit.aes = FALSE)
g_dots

ggsave(g_dots, file = "../Figures/Ttn_dots.pdf", height = 7, width = 6)
ggsave(g_dots, file = "../Figures/Ttn_dots.png", height = 7, width = 6)

g_dots_box_base = g_dots + 
  add_scalebar(dist_um = 50,
               x = range_x[2], y = -range_y[2] - 0.4,
               box.col = "white")
g_dots_box = g_dots_box_base + geom_polygon(aes(x = segmentation_vertices_x_global_affine,
                                                y = -segmentation_vertices_y_global_affine),
                                            data = data.frame(
                                              segmentation_vertices_x_global_affine = c(range_x_zoom[c(1,1,2,2)]), 
                                              segmentation_vertices_y_global_affine = -c(range_y_zoom[c(1,2,2,1)])),
                                            inherit.aes = FALSE, fill = NA, colour = "white", size = 2) +
  NULL
g_dots_box

ggsave(g_dots_box, file = "../Figures/Ttn_dots_box.pdf", height = 7, width = 6)
ggsave(g_dots_box, file = "../Figures/Ttn_dots_box.png", height = 7, width = 6)

g_dots_zoom = g_dots_box_base + xlim(range_x_zoom) + ylim(range_y_zoom) + 
  add_scalebar(dist_um = 50,
               x = range_x_zoom[2] - 0.1*diff(range_x_zoom),
               y = -range_y_zoom[2] + 0.1*diff(range_y_zoom),
               box.col = "white")

g_dots_zoom 

ggsave(g_dots_zoom, file = "../Figures/Ttn_dots_zoom.pdf", height = 7/2.4, width = 6/2.4)
ggsave(g_dots_zoom, file = "../Figures/Ttn_dots_zoom.png", height = 7/2.4, width = 6/2.4)

# transparent labels
gene_labels = data.frame(gene = c("Ttn", "Tbx5", "Cdh5" , "Dlk1"),
                         colour = c("red",  "cyan", "green","orange"),
                         x = 1:4,
                         y = 1)

g = ggplot(gene_labels, aes(x = x, y = y)) + 
  geom_text(aes(label = gene, colour = colour), fontface = "italic") + 
  scale_colour_identity() + 
  theme_classic() + 
  theme(plot.background = element_rect(fill = "transparent")) +
  theme(panel.background = element_rect(fill = "transparent")) +
  theme(axis.line = element_blank()) + 
  theme(axis.text = element_blank()) +
  theme(axis.ticks = element_blank()) +
  xlab("") +
  ylab("")
g
ggsave(g, file = "../Figures/Ttn_dots_legend.pdf", height = 0.5, width = 2)
```

# Subcellular localisation of Gbx2 and Otx2

```{r}
cell_info = rbind(read.delim("../analysis_output/alsu/meta_section_MHB_embryo2z2.tab",
                             header = TRUE, row.names = 1),
                  read.delim("../analysis_output/alsu/meta_section_MHB_embryo2z5.tab",
                             header = TRUE, row.names = 1))
mhb_cells = rownames(cell_info)

boundary_polygons = getSegmentationVerticesDF(colData(sce)[mhb_cells,],
                                              xname = "segmentation_vertices_x_global_affine",
                                              yname = "segmentation_vertices_y_global_affine",
                                              othercols = c("uniqueID","z","celltype_mapped_refined","cluster","cluster_sub","embryo"))

colour_scheme = c("gray75","cornflowerblue","black")

mRNA_Ttn = mRNA_df[mRNA_df$z == 2 & mRNA_df$embryo == "embryo2",]

boundary_polygons_sub = boundary_polygons

boundary_polygons_sub$brain_region <- cell_info[as.character(boundary_polygons$uniqueID), "subgroup"]

range_x = range(boundary_polygons_sub$segmentation_vertices_x_global_affine)
range_y = range(-boundary_polygons_sub$segmentation_vertices_y_global_affine)

boundary_polygons_sub$Ttn <- logcounts(sce)["Ttn", boundary_polygons_sub$uniqueID]

boundary_polygons_sub$Ttn_norm <- unsplit(tapply(boundary_polygons_sub$Ttn,
                                                 boundary_polygons_sub$embryo,
                                                 function(x) x/max(x)),
                                          boundary_polygons_sub$embryo)

# with black background
g = ggplot(subset(boundary_polygons_sub, embryo == "embryo2" & z == 2),
           aes(x = segmentation_vertices_x_global_affine,
               y = -segmentation_vertices_y_global_affine,
               fill = Ttn_norm)) + 
  geom_polygon(fill = "black", colour = "white", size = 0.1, aes(group = uniqueID)) +
  theme_classic() +
  scale_fill_gradient2(low = colour_scheme[1], mid = colour_scheme[2], high = colour_scheme[3]) +
  guides(fill = guide_legend(title = "Ttn normalised logcounts")) +
  xlab("") +
  ylab("") +
  
  xlim(range_x) +
  ylim(range_y) +
  
  theme(panel.background = element_rect(fill = "black")) +
  theme(plot.background = element_rect(fill = "black")) +
  theme(axis.text = element_blank()) + 
  theme(axis.ticks = element_blank()) +
  coord_fixed() +
  NULL

g

x_range = range(g$data$segmentation_vertices_x_global_affine)
y_range = range(g$data$segmentation_vertices_y_global_affine)

g_dots = g + 
  geom_point(aes(x = x_global_affine, y = -y_global_affine), 
             data = subset(mRNA_Ttn, embryo == "embryo2" & z == 2 & geneID == "Gbx2"),
             size = 0.75,
             alpha = 0.7,
             colour = "purple",
             shape = 16,
             inherit.aes = FALSE) + 
  geom_point(aes(x = x_global_affine, y = -y_global_affine), 
             data = subset(mRNA_Ttn, embryo == "embryo2" & z == 2 & geneID == "Otx2"),
             size = 0.75,
             alpha = 0.7,
             colour = "orange",
             shape = 16,
             inherit.aes = FALSE) + 
  add_scalebar(dist_um = 50, x = x_range[1] + 0.3*diff(x_range),
               y = -mean(y_range[1]) - 0.05, box.col = "white") +
  NULL
g_dots
ggsave(g_dots, file = "../Figures/Gbx2_Otx2_dots.pdf", height = 2.5, width = 3.5)
ggsave(g_dots, file = "../Figures/Gbx2_Otx2_dots.png", height = 2.5, width = 3.5)

saveRDS(c(x = x_range[1] + 0.3*diff(x_range),
          y = -mean(y_range[1]) - 0.05),
        file = "../analysis_output/E8.5/MHB_scalebar_location.Rds")

# add dotted line for selected regions

MHB_coords_mid_ind = chull(subset(boundary_polygons_sub, brain_region != "Hindbrain")[, c("segmentation_vertices_x_global_affine", "segmentation_vertices_y_global_affine")])
MHB_coords_mid = subset(boundary_polygons_sub, brain_region != "Hindbrain")[MHB_coords_mid_ind,]

MHB_coords_hind_ind = chull(subset(boundary_polygons_sub, brain_region == "Hindbrain")[, c("segmentation_vertices_x_global_affine", "segmentation_vertices_y_global_affine")])
MHB_coords_hind = subset(boundary_polygons_sub, brain_region == "Hindbrain")[MHB_coords_hind_ind,]


g_dots + geom_polygon(aes(x = segmentation_vertices_x_global_affine, 
                          y = -segmentation_vertices_y_global_affine),
                      data = MHB_coords_mid, inherit.aes = FALSE,
                      fill = NA, colour = "orange", linetype = "dashed", size = 1) + 
  geom_polygon(aes(x = segmentation_vertices_x_global_affine, 
                   y = -segmentation_vertices_y_global_affine),
               data = MHB_coords_hind, inherit.aes = FALSE,
               fill = NA, colour = "purple", linetype = "dashed", size = 1)



# altogether to reorder
g = ggplot(subset(boundary_polygons_sub, embryo == "embryo2" & z == 2),
           aes(x = segmentation_vertices_x_global_affine,
               y = -segmentation_vertices_y_global_affine,
               group = uniqueID,
               fill = Ttn_norm)) + 
  theme_classic() +
  scale_fill_gradient2(low = colour_scheme[1], mid = colour_scheme[2], high = colour_scheme[3]) +
  guides(fill = guide_legend(title = "Ttn normalised logcounts")) +
  xlab("") +
  ylab("") +
  
  xlim(range_x) +
  ylim(range_y) +
  
  theme(panel.background = element_rect(fill = "black")) +
  theme(plot.background = element_rect(fill = "black")) +
  theme(axis.text = element_blank()) + 
  theme(axis.ticks = element_blank()) +
  coord_fixed() +
  
  geom_polygon(aes(x = segmentation_vertices_x_global_affine,
                   y = -segmentation_vertices_y_global_affine),
               data = MHB_coords_mid, inherit.aes = FALSE,
               fill = NA, colour = "orange", linetype = "dotted", size = 0.5) +
  geom_polygon(aes(x = segmentation_vertices_x_global_affine,
                   y = -segmentation_vertices_y_global_affine),
               data = MHB_coords_hind, inherit.aes = FALSE,
               fill = NA, colour = "purple", linetype = "dashed", size = 0.5) +
  
  
  # cell polygons
  geom_polygon(fill = NA, aes(colour = brain_region), size = 0.1, show.legend = FALSE) +
  
  scale_colour_manual(values = c("Midbrain" = "orange", "Forebrain" = "orange",
                                 "Hindbrain" = "purple")) +
  
  
  geom_point(aes(x = x_global_affine, y = -y_global_affine), 
             data = subset(mRNA_Ttn, embryo == "embryo2" & z == 2 & geneID == "Gbx2"),
             size = 0.75,
             alpha = 0.7,
             colour = "purple",
             shape = 16,
             inherit.aes = FALSE) + 
  geom_point(aes(x = x_global_affine, y = -y_global_affine), 
             data = subset(mRNA_Ttn, embryo == "embryo2" & z == 2 & geneID == "Otx2"),
             size = 0.75,
             alpha = 0.7,
             colour = "orange",
             shape = 16,
             inherit.aes = FALSE) + 
  
  
  NULL


g

ggsave(g, file = "../Figures/Gbx2_Otx2_dots_borders.pdf", height = 2.5, width = 3.5)
ggsave(g, file = "../Figures/Gbx2_Otx2_dots_borders.png", height = 2.5, width = 3.5)

```



# Nowotschin joint UMAP

colours copied from the Nowotschin report

```{r}
# this is done manually, by looking at the Extended Data Figure 
clusterGroups = c(
  "10" = "Thyroid",
  "0" = "Thymus",
  "7" = "Lung 1",
  "8" = "Lung 2",
  "9" = "Liver",
  "3" = "Pancreas 1",
  "4" = "Pancreas 2",
  "2" = "Small intestine",
  "6" = "Large intestine/Colon"
)

# viridis hard to distinguish
AP_pseudo_colours = RColorBrewer::brewer.pal(length(clusterGroups), "Set3")
AP_pseudo_colours[length(clusterGroups)] = RColorBrewer::brewer.pal(length(clusterGroups) + 1, "Set3")[length(clusterGroups) + 1]
names(AP_pseudo_colours) <- clusterGroups
plot(1:length(clusterGroups), col = AP_pseudo_colours, pch = 16, cex = 5)
```

```{r}
now_joint_df = readRDS("../analysis_output/E8.5/E8.5_Nowotschin_joint_mnn_df.Rds")
now_joint_df$embryo = as.character(substring(now_joint_df$split, 1, 7))
now_joint_df$embryo[now_joint_df$embryo == "Nowotsc"] <- "Nowotschin"
now_joint_df$embryo <- factor(now_joint_df$embryo)
```

```{r}
g = ggplot(now_joint_df[sample(rownames(now_joint_df)), ], aes(x = umap1, y = -umap2)) +
  geom_point(aes(colour = type_lab), size = 0.5, alpha = 0.3) + 
  scale_colour_manual(values = c("seqFISH" = "navy", "Nowotschin" = "orange", "Nowotschin_other" = "yellow4"),
                      labels = c("seqFISH" = "seqFISH", 
                                 "Nowotschin" = "Nowotschin et al (Gut tube)", 
                                 "Nowotschin_other" = "Nowotschin et al (other)")) +
  theme_classic() +
  theme(legend.position = "bottom") +
  theme(axis.text = element_blank()) + 
  theme(axis.ticks = element_blank()) +
  coord_fixed() +
  xlab("UMAP1") + 
  ylab("UMAP2") +
  guides(colour = guide_legend(title = "",
                               override.aes = list(size = 8, alpha = 1))) +
  NULL
g
ggsave(g, file = "../Figures/joint_umap_type_Nowotschin.pdf",
       height = 5, width = 5)

gg = g + facet_wrap(~embryo)
gg
ggsave(gg, file = "../Figures/joint_umap_embryo_split_Nowotschin.pdf",
       height = 6, width = 5)

```

# Calculate principal curve of the seqFISH nowotschin cells

so that i can then overlay an A-P axis line

```{r}
DV_class = readRDS("../analysis_output/E8.5/gut_tube_DV_split.Rds")

D_cells = DV_class[["DV_class"]]
A_vals = DV_class[["AP_value"]]

now_joint_df[, "DV_class"] <- D_cells[rownames(now_joint_df)]
now_joint_df[, "AP_value"] <- A_vals[rownames(now_joint_df)]

now_joint_df_sub = subset(now_joint_df, 
                          TRUE
)
dim(now_joint_df_sub)

table(now_joint_df_sub$DV_class, useNA = "always")

plot(now_joint_df_sub$umap1, -now_joint_df_sub$umap2, col = factor(now_joint_df_sub$DV_class),
     asp = 1,
     pch = 16,
     cex = 0.75)

g = ggplot(now_joint_df_sub, aes(x = umap1, y = -umap2)) + 
  geom_point(colour = "lightgrey", size = 0.5, alpha = 0.2) +
  geom_point(aes(colour = DV_class), size = 0.5,
             data = subset(now_joint_df_sub, !is.na(DV_class)), alpha = 0.5) + 
  scale_colour_manual(values = c("Dorsal" = "orange", "Ventral" = "purple")) + 
  theme_classic() + 
  theme(axis.ticks = element_blank()) + 
  theme(axis.text = element_blank()) + 
  xlab("UMAP1") + ylab("UMAP2") +
  theme(legend.position = "none") +
  coord_fixed() + 
  NULL
g
ggsave(g, file = "../Figures/joint_umap_nowotschin_DV_coloured.pdf",
       height = 5, width = 5)

gList = list()

embs = c("embryo1", "embryo2", "embryo3")
dvs = c("Ventral", "Dorsal")

for (emb in embs) {
  for (dv in dvs) {
    
    g = ggplot(now_joint_df_sub, aes(x = umap1, y = -umap2)) + 
      geom_point(colour = "lightgrey", size = 1) + 
      geom_point(data = subset(now_joint_df_sub, embryo == emb & DV_class == dv),
                 aes(colour = rank(AP_value)), size = 1) +
      scale_colour_gradient2(low = "yellow", mid = "purple", high = "black",
                             midpoint = 0.5*length(subset(now_joint_df_sub, embryo == emb & DV_class == dv)$AP_value)) + 
      theme_classic() + 
      theme(axis.ticks = element_blank()) + 
      theme(axis.text = element_blank()) + 
      xlab("UMAP1") + ylab("UMAP2") +
      theme(legend.position = "none") +
      ggtitle(paste0(emb, ": ", dv)) +
      coord_fixed() + 
      NULL
    
    gList[[paste0(emb, "_", dv)]] <- g
  }
}

g = wrap_plots(gList, nrow = 2, ncol = 3, byrow = FALSE)

ggsave(g, file = "../Figures/joint_umap_nowotschin_AP_coloured_split.pdf",
       height = 10, width = 15)

# also make the legend
df = data.frame(x = 1:10)
g = ggplot(df, aes(x = x, y = x)) + 
  geom_point(aes(colour = x)) + 
  scale_colour_gradient2(low = "yellow", mid = "purple", high = "black",
                         midpoint = 5,
                         labels = c("Anterior", "Posterior"), 
                         breaks = c(1,10),
                         limits = c(1,10)) + 
  guides(colour = guide_colourbar(title = "")) + 
  theme(legend.position = "bottom")
g

gg = as_ggplot(get_legend(g))
ggsave(gg, file = "../Figures/AP_legend.pdf", height = 1, width = 2)

```

# Perform clustering of Nowotschin and seqFISH cells

and select only those clusters with a high number of nowotschin cells
and with a high proportion of gut tube related cells

```{r}
now_pca = readRDS("../analysis_output/E8.5/nowotschin_joint_mnn_pca.Rds")
dim(now_pca)

graph = buildSNNGraph(now_pca, transposed = TRUE)
clusters_graph = cluster_louvain(graph)
clusters = as.numeric(membership(clusters_graph))

names(clusters) <- rownames(now_pca)
table(clusters, useNA = "always")

now_joint_df$joint_cluster = clusters

table(now_joint_df$joint_cluster, now_joint_df$type)

# which clusters have at least 100 nowotschin cells
# where at least half of them are gut tube types

out = list()
cluster_val = 1
for (cluster_val in 1:length(unique(clusters))) {
  ind = now_joint_df$joint_cluster == cluster_val & now_joint_df$type == "Nowotschin"
  sumind = sum(ind)
  meanind = mean(now_joint_df$type_lab[ind] == "Nowotschin")
  out[[cluster_val]] <- c(sumind, meanind)
}
out_df = do.call(rbind, out)
plot(out_df, type = "n")
text(out_df, labels = 1:nrow(out_df))

selectedClusters = which(out_df[,2] > 0.5)
selectedClusters

g = ggplot(now_joint_df[sample(rownames(now_joint_df)), ], aes(x = umap1, y = -umap2)) +
  geom_point(aes(colour = factor(joint_cluster %in% selectedClusters)), size = 0.5, alpha = 0.3) + 
  # scale_colour_manual(values = c("seqFISH" = "navy", "Nowotschin" = "orange", "Nowotschin_other" = "yellow4"),
  #                     labels = c("seqFISH" = "seqFISH", 
  #                                "Nowotschin" = "Nowotschin et al (Gut tube)", 
  #                                "Nowotschin_other" = "Nowotschin et al (other)")) +
  theme_classic() +
  theme(legend.position = "bottom") +
  # coord_fixed() +
  theme(axis.text = element_blank()) + 
  theme(axis.ticks = element_blank()) +
  # facet_wrap(~type) +
  coord_fixed() +
  xlab("UMAP1") + 
  ylab("UMAP2") +
  guides(colour = guide_legend(title = "",
                               override.aes = list(size = 8, alpha = 1))) +
  NULL
g

# now_joint_df$ctype_mask_cluster = now_joint_df$ctype
# now_joint_df$ctype_mask_cluster[now_joint_df$type == "seqFISH" & 
#                                   (
#                                   # (!now_joint_df$joint_cluster %in% selectedClusters |
#                                   !now_joint_df$celltype %in% c("Gut tube", "Surface ectoderm"))] <- NA
# now_joint_df$ctype_mask_cluster <- factor(now_joint_df$ctype_mask_cluster,
#                                           levels = names(AP_pseudo_colours))
# 
# table(now_joint_df$ctype, now_joint_df$ctype_mask_cluster, useNA = "always")
# 
# table(now_joint_df$celltype, !is.na(now_joint_df$ctype_mask_cluster), useNA = "always")
```


```{r}
g = ggplot(now_joint_df[sample(rownames(now_joint_df)), ], aes(x = umap1, y = -umap2)) +
  geom_point(aes(colour = ctype_mask), size = 0.5, alpha = 0.3) + 
  # scale_colour_manual(values = c("seqFISH" = "navy", "Nowotschin" = "orange", "Nowotschin_other" = "yellow4"),
  #                     labels = c("seqFISH" = "seqFISH", 
  #                                "Nowotschin" = "Nowotschin et al (Gut tube)", 
  #                                "Nowotschin_other" = "Nowotschin et al (other)")) +
  scale_colour_manual(values = AP_pseudo_colours, na.value = "lightgrey") + 
  theme_classic() +
  theme(legend.position = "bottom") +
  # coord_fixed() +
  theme(axis.text = element_blank()) + 
  theme(axis.ticks = element_blank()) +
  facet_wrap(~type) +
  coord_fixed() +
  xlab("UMAP1") + 
  ylab("UMAP2") +
  guides(colour = guide_legend(title = "",
                               override.aes = list(size = 8, alpha = 1))) +
  NULL
g
ggsave(g, file = "../Figures/joint_umap_celltype_split_Nowotschin.pdf",
       height = 5, width = 9)
```

```{r}
g = ggplot(now_joint_df[sample(rownames(now_joint_df)), ], aes(x = umap1, y = -umap2)) +
  geom_point(aes(colour = celltype), size = 0.5, alpha = 0.3) + 
  scale_colour_manual(values = celltype_colours, na.value = "lightgrey") + 
  theme_classic() +
  theme(legend.position = "none") +
  theme(axis.text = element_blank()) + 
  theme(axis.ticks = element_blank()) +
  facet_wrap(~type) +
  coord_fixed() +
  xlab("UMAP1") + 
  ylab("UMAP2") +
  guides(colour = guide_legend(title = "",
                               override.aes = list(size = 8, alpha = 1))) +
  NULL
g
ggsave(g, file = "../Figures/joint_umap_celltype_atlas_split_Nowotschin.pdf",
       height = 5, width = 9)

```

# cell cell contact for Nowotschin et al data

```{r}
sce$gut_tube_subtype <- as.character(now_joint_df[colnames(sce), "ctype_mask"])

sce_sub = sce[, !is.na(sce$gut_tube_subtype) & sce$celltype_mapped_refined == "Gut tube"]

graph = readRDS("../analysis_output/E8.5/E8.5_neighbourGraph_1.3.Rds")

nperm = 500

out_all = cellCellContact(sce_sub, group = "gut_tube_subtype", graph, nperm = nperm)
g = cellCellContactMap(out_all, order = rev(names(AP_pseudo_colours)))
g

split_categories = interaction(sce_sub$embryo, sce_sub$z)
sce_split = splitSCE(sce_sub, split_categories)

outList = lapply(sce_split, cellCellContact, group = "gut_tube_subtype",
                 graph = graph, nperm = nperm, plot = FALSE)
gList = lapply(outList, cellCellContactMap, order = rev(names(AP_pseudo_colours)))
gList2 = sapply(names(gList), function(n) gList[[n]] + ggtitle(n), simplify = FALSE)
wrap_plots(gList2)

gList[[1]]

pmat_all = lapply(outList, "[[", 2)
all_celltypes = sort(unique(unlist(lapply(pmat_all,rownames))))
pmatArray = abind::abind(lapply(pmat_all, function(x) {
  x_all = matrix(NA, nrow = length(all_celltypes),
                 ncol = length(all_celltypes), dimnames = list(all_celltypes, all_celltypes))
  x_all[rownames(x), colnames(x)] <- x
  return(x_all)
}), along = 3)
pmatMean = apply(pmatArray, 1:2, mean, na.rm = TRUE)
g = cellCellContactMap(list(obs = pmatMean, pmat = pmatMean),
                       order = rev(names(AP_pseudo_colours)))
g

h = cellCellContactHeatmapTriangle(list(obs = pmatMean, pmat = pmatMean),
                                   col_ann = AP_pseudo_colours,
                                   order = setdiff(names(AP_pseudo_colours), "Large intestine/Colon"),
                                   exclude = "Large intestine/Colon")
print(h)

pdf("../Figures/neighbourhood_Nowotschin.pdf",
    height = 6, width = 6)
print(h)
dev.off()

h = cellCellContactHeatmapTriangle(list(obs = pmatMean, pmat = pmatMean),
                                   col_ann = AP_pseudo_colours,
                                   # order = setdiff(names(AP_pseudo_colours), "Large intestine/Colon"),
                                   exclude = "Large intestine/Colon",
                                   split_n = 5)
print(h)

pdf("../Figures/neighbourhood_Nowotschin_clustered.pdf",
    height = 6, width = 6)
print(h)
dev.off()
```

# cell cell separability

```{r}
dist_sep = cellCellSeparability(sce_sub, split = interaction(sce_sub$embryo, sce_sub$z),
                                group = "gut_tube_subtype",
                                coordNames = c("x_global_affine", 
                                               "y_global_affine"),
                                plot = FALSE)

g = cellCellSeparabilityMap(dist_sep, order = rev(names(AP_pseudo_colours)))
g
ggsave(g, file = "../Figures/separability_Nowotschin.pdf",
       height = 8, width = 8)
```

# Figure of nowotschin cells in space

```{r}
g = ggplot(subset(now_joint_df, type == "seqFISH"), aes(x = coord_x, y = coord_y)) +
  geom_point(aes(colour = ctype_mask), size = 0.2) +
  scale_colour_manual(values = AP_pseudo_colours, na.value = "lightgrey") +
  theme_classic() +
  theme(legend.position = "right") +
  coord_fixed() +
  facet_wrap(split~., dir = "v", ncol = 3) +
  NULL
g 

now_joint_df$ctype_mask_gut = now_joint_df$ctype_mask
now_joint_df$ctype_mask_gut[now_joint_df$celltype == "Surface ectoderm"] <- NA
now_joint_df$ctype_mask_gut = factor(now_joint_df$ctype_mask_gut, levels = 
                                       names(AP_pseudo_colours))

g = ggplot(subset(now_joint_df, type == "seqFISH"), aes(x = coord_x, y = coord_y)) +
  geom_point(aes(colour = ctype_mask_gut), size = 0.2) +
  scale_colour_manual(values = AP_pseudo_colours, na.value = "lightgrey") +
  theme_classic() +
  theme(legend.position = "right") +
  coord_fixed() +
  facet_wrap(split~., dir = "v", ncol = 3) +
  guides(colour = guide_legend(title = "", override.aes = list(size = 5, alpha = 1))) + 
  NULL
g 
ggsave(g, file = "../Figures/gut_tube_coloured_points_nowotschin.pdf", height = 10, width = 20)
```

```{r}
g = ggplot(now_joint_df[sample(rownames(now_joint_df)), ], aes(x = umap1, y = -umap2)) +
  geom_point(aes(colour = ctype_mask_gut), size = 0.5, alpha = 0.3) + 
  scale_colour_manual(values = AP_pseudo_colours, na.value = "lightgrey") + 
  theme_classic() +
  theme(legend.position = "bottom") +
  theme(axis.text = element_blank()) + 
  theme(axis.ticks = element_blank()) +
  facet_wrap(~type) +
  coord_fixed() +
  xlab("UMAP1") + 
  ylab("UMAP2") +
  guides(colour = guide_legend(title = "",
                               override.aes = list(size = 8, alpha = 1))) +
  NULL
g
ggsave(g, file = "../Figures/joint_umap_celltype_split_gut_tube_Nowotschin.pdf",
       height = 5, width = 9)
```

# polygon graph of with nowotschin

```{r}
celltype_mapped_col <- AP_pseudo_colours[as.character(now_joint_df$ctype_mask_gut)]
names(celltype_mapped_col) <- rownames(now_joint_df)

boundary_polygons = getSegmentationVerticesDF(colData(sce),
                                              xname = "segmentation_vertices_x_global_affine",
                                              yname = "segmentation_vertices_y_global_affine",
                                              othercols = c("uniqueID","z","embryo"))

embryovals = unique(boundary_polygons$embryo)
zvals = unique(boundary_polygons$z)

zoomvals = list("embryo1" = c(x = -1.7, y = -2.2),
                "embryo2" = c(x = 0, y = -2.7),
                "embryo3" = c(x = -1.2, y = -3.1))

for (embryoval in embryovals) {
  for (zval in zvals) {
    
    g = ggplot(subset(boundary_polygons, embryo == embryoval & z == zval)) + 
      geom_polygon(aes(group = uniqueID,
                       x = segmentation_vertices_x_global_affine,
                       y = -segmentation_vertices_y_global_affine,
                       fill = uniqueID), colour = "black",
                   size = 0.1) + 
      scale_fill_manual(values = celltype_mapped_col, na.value = "lightgrey") +
      theme_classic() +
      coord_fixed() +
      theme(legend.position = "none") +
      NULL
    g
    
    gg = g + theme(axis.line = element_blank()) + 
      theme(axis.text = element_blank()) +
      theme(axis.ticks = element_blank()) + 
      theme(axis.title = element_blank()) +
      add_scalebar() + 
      NULL
    gg
    ggsave(gg, file = paste0("../Figures/gut_tube_coloured_",
                             embryoval, "_", zval, ".png"),
           height = 10, width = 8 )
    
    gg = g + theme(axis.line = element_blank()) + 
      theme(axis.text = element_blank()) +
      theme(axis.ticks = element_blank()) + 
      theme(axis.title = element_blank()) +
      add_scalebar(dist_um = 50, x = zoomvals[[embryoval]]["x"], y = zoomvals[[embryoval]]["y"]) +
      NULL
    gg
    ggsave(gg, file = paste0("../Figures/gut_tube_coloured_forzoom_",
                             embryoval, "_", zval, ".png"),
           height = 10, width = 8 )
  }
}
```

# polygon graph of Nowotschin, inc surface ectoderm

```{r}
celltype_mapped_col <- AP_pseudo_colours[as.character(now_joint_df$ctype_mask)]
names(celltype_mapped_col) <- rownames(now_joint_df)

boundary_polygons = getSegmentationVerticesDF(colData(sce),
                                              xname = "segmentation_vertices_x_global_affine",
                                              yname = "segmentation_vertices_y_global_affine",
                                              othercols = c("uniqueID","z","embryo"))

embryovals = unique(boundary_polygons$embryo)
zvals = unique(boundary_polygons$z)

for (embryoval in embryovals) {
  for (zval in zvals) {
    
    g = ggplot(subset(boundary_polygons, embryo == embryoval & z == zval)) + 
      geom_polygon(aes(group = uniqueID,
                       x = segmentation_vertices_x_global_affine,
                       y = -segmentation_vertices_y_global_affine,
                       fill = uniqueID), colour = "black",
                   size = 0.1) + 
      scale_fill_manual(values = celltype_mapped_col, na.value = "lightgrey") +
      theme_classic() +
      coord_fixed() +
      theme(legend.position = "none") +
      NULL
    g
    
    gg = g + theme(axis.line = element_blank()) + 
      theme(axis.text = element_blank()) +
      theme(axis.ticks = element_blank()) + 
      theme(axis.title = element_blank()) +
      NULL
    gg
    ggsave(gg, file = paste0("../Figures/gut_tube_coloured_incSE_",
                             embryoval, "_", zval, ".png"),
           height = 10, width = 8 )
  }
}
```

# Perform differential expression of these subtype groups

Do it while including surface ectoderm and without

```{r}
sce_sub = sce[, !is.na(sce$gut_tube_subtype) & sce$celltype_mapped_refined == "Gut tube"]

marker_array = getMarkerArray(sce_sub, 
                              group = "gut_tube_subtype",
                              subset = sce_sub$gut_tube_subtype != "Large intestine/Colon",
                              block = "embryo",
                              pseudobulk = TRUE)

discriminating_array_list = list()

groupValue = "gut_tube_subtype"

groups = setdiff(sort(unique(colData(sce)[, groupValue])), "Large intestine/Colon")

for (grouplevel in groups) {
  
  print(grouplevel)
  
  groups2 = setdiff(groups, grouplevel)
  
  for (grouplevel2 in groups2) {
    
    print(grouplevel2)
    
    ma = getMarkerArray(sce_sub,
                        group = groupValue,
                        block = "embryo",
                        subset = colData(sce_sub)[, groupValue] %in% c(grouplevel,
                                                                       grouplevel2),
                        pseudobulk = TRUE,
                        verbose = FALSE)
    
    discriminating_array_list[[grouplevel]][[grouplevel2]] <- ma
    
  }
  
}

discriminating_array_list[["Lung 1"]][["Lung 2"]]["Shh",,]
marker_array["Shh",,]

lung_comparison_tab = discriminating_array_list[["Lung 1"]][["Lung 2"]][,"Lung 1",]
colnames(lung_comparison_tab) <- c(
  "p.value", "FDR", "LFC", "meanExpression.Lung1", "meanExpression.Lung2"
)

# output res as an excel spreadsheet, each cell type is a table
res_table_file = "../analysis_output/E8.5/Nowotschin_DE_Lung1_Lung2.xlsx"

wb = createWorkbook()
sh =  "Lung 1 vs Lung 2"
addWorksheet(wb,sheetName = sh)
writeData(wb, 
          sh, 
          lung_comparison_tab,
          rowNames = TRUE)
saveWorkbook(wb,
             file = res_table_file,
             overwrite = TRUE)


markerArrayPlot(marker_array,
                grouplevel = "Lung 1",
                otherGroupLabel = "All other cells",
                onlyLabelTopRanked = TRUE,
                FDR = 0.2,
                LFC = 0.1)

saveRDS(discriminating_array_list, file = "../analysis_output/E8.5/gut_tube_discriminating_array_list.Rds")
saveRDS(marker_array, file = "../analysis_output/E8.5/gut_tube_marker_array.Rds")
```

# Expression plot of Sox2 and Nkx2-1 (imputed) for gut tube only

```{r}
for (gene in c("Sox2", "Osr1", "Smoc2", "Tbx1")) {
  print(gene)
  
  assay(sce, "mask") <- assay(sce, "logcounts")
  assay(sce, "mask")[gene, is.na(sce$gut_tube_subtype) | sce$celltype_mapped_refined != "Gut tube"] <- NA
  
  for (embryoval in c("embryo1", "embryo2", "embryo3")) {
    
    g = plotEmbryo(sce, 
                   gene_name = gene, 
                   assayName = "mask",
                   quantileNorm = FALSE,
                   embryo_subset = embryoval,
                   z_subset = "2",
                   outline = TRUE) + 
      add_scalebar(dist_um = 50, x = zoomvals[[embryoval]]["x"],
                   y = zoomvals[[embryoval]]["y"],
                   box.col = "black",
                   border.size = 5)
    # g + ggtitle(gene)
    ggsave(g, file = paste0("../Figures/corrected_gut_tube_", gene, "_", embryoval, "_z2.pdf"),
           height = 20, width = 24)
    
  }
}

g = plotEmbryo(sce, 
               gene_name = gene, 
               assayName = "mask",
               quantileNorm = FALSE,
               embryo_subset = c("embryo1", "embryo2", "embryo3"),
               z_subset = c("2","5"),
               outline = TRUE)
g
ggsave(g, file = paste0("../Figures/corrected_gut_tube_", gene, ".pdf"),
       height = 30, width = 40)
```

Full embryo digital in situ for the genes Evan imaged with HCR

```{r}
genesList = list(c("Tbx1", "Shh"),
                 c("Smoc2", "Tbx3"),
                 c("Smoc2", "Gata3"),
                 c("Gbx2", "Otx2"))

for (i in 1:length(genesList)) {
  
  genes = genesList[[i]]
  if (identical(genes, c("Gbx2", "Otx2"))) {
    genes_cols = c("purple", "orange")
  } else {
    genes_cols = c("red", "cyan")
  }
  names(genes_cols) = genes
  
  for (embryoval in unique(sce$embryo)) {
    for (zval in unique(sce$z)) {
      
      dots_df = subset(mRNA_df, geneID %in% genes & 
                         z == zval &
                         embryo == embryoval)
      
      g = plotEmbryo(sce, 
                     gene_name = rownames(sce)[1], 
                     assayName = "logcounts",
                     quantileNorm = FALSE,
                     embryo_subset = embryoval,
                     z_subset = zval,
                     outline = FALSE,
                     colours = rep("grey12", 3)) + 
        add_scalebar(box.col = "white", border.size = 10)
      
      gg = g + 
        geom_scattermore(aes(x = x_global_affine, y = -y_global_affine, colour = geneID),
                         data = dots_df,
                         pointsize = 3.3,
                         pixels = c(1024,1024),
                         alpha = 0.5) + 
        scale_colour_manual(values = genes_cols) + 
        theme(plot.background = element_rect(fill = "black"),
              panel.background = element_rect(fill = "black")) + 
        NULL
      
      ggsave(gg, file = paste0("../Figures/full_digitalinsitu_",
                               paste0(genes, collapse = "_"),
                               "_",embryoval, "_", zval,
                               ".pdf"),
             height = 20, width = 24)
    }
  }
}
```

Full embryo digital in situ for Ttn

```{r}
genes = c("Ttn")
genes_cols = c("red")

names(genes_cols) = genes
plot(1:length(genes), pch = 16, col = genes_cols, cex = 3)

embryoval = "embryo1"
zval = "2"

dots_df = subset(mRNA_df, geneID %in% genes & 
                   z == zval &
                   embryo == embryoval)

g = plotEmbryo(sce, 
               gene_name = rownames(sce)[1], 
               assayName = "logcounts",
               quantileNorm = FALSE,
               embryo_subset = embryoval,
               z_subset = zval,
               outline = FALSE,
               colours = rep("grey12", 3)) + 
  add_scalebar(box.col = "white")

gg = g + 
  geom_scattermore(aes(x = x_global_affine, y = -y_global_affine, colour = geneID),
                   data = dots_df,
                   pointsize = 1.1,
                   pixels = c(1024,1024)) + 
  scale_colour_manual(values = genes_cols) + 
  theme(plot.background = element_rect(fill = "black"),
        panel.background = element_rect(fill = "black")) + 
  NULL

ggsave(gg, file = paste0("../Figures/full_digitalinsitu_",
                         paste0(genes, collapse = "_"),
                         "_",embryoval, "_", zval,
                         ".pdf"),
       height = 20, width = 24)
```

Expression of genes within a celltype

```{r}
unique_tstats = readRDS("../analysis_output/E8.5/unique_tstats.Rds")
types = c(1:2)

for (type in types) {
  if (type == 1) {
    genes = names(unique_tstats[["Forebrain/Midbrain/Hindbrain"]])
    ctype = "Forebrain/Midbrain/Hindbrain"
  } 
  if (type == 2) {
    genes = c("En1", "Lhx2", "Hoxb1")
    ctype = "Forebrain/Midbrain/Hindbrain"
  }
  embryoval = "embryo1"
  zval = "2"
  
  for (gene in genes) {  
    print(gene)
    
    assay(sce, "fake") <- assay(sce, "logcounts")
    assay(sce, "fake")[gene, sce$celltype_mapped_refined != ctype] <- NA
    
    g = plotEmbryo(sce, 
                   gene_name = gene, 
                   assayName = "fake",
                   quantileNorm = FALSE,
                   embryo_subset = embryoval,
                   z_subset = zval,
                   outline = "grey80"#,
    ) + 
      add_scalebar(box.col = "white")
    
    gg = g + add_scalebar(x = -1, y = 3.2) + ggtitle(gene) + 
      theme(plot.title = element_text(face = "italic", hjust = 0.25, size = 20))
    
    gg
    
    if (type == 2) {
      ggsave(gg, file = paste0("../Figures/",
                               gsub("/", "_", ctype), "_expression_",
                               paste0(gene, collapse = "_"),
                               "_",embryoval, "_", zval,
                               ".pdf"),
             height = 10, width = 6)
    }
    
    if (type == 1) { 
      ggsave(gg, file = paste0("../Figures/",
                               gsub("/", "_", ctype), "_expression_",
                               paste0(gene, collapse = "_"),
                               "_rank_",
                               which(genes == gene),"_",
                               embryoval, "_", zval,
                               ".pdf"),
             height = 10, width = 6)
    }
  }
}
```

# Plot of the gut tube cells

```{r}
cells = as.character(unlist(sapply(list.files("../analysis_output/guttubecells/", pattern = ".Rds",
                                              full.names = TRUE), 
                                   readRDS)))

sce$gut_tube = ifelse(colnames(sce) %in% cells, 1, 0)
assay(sce, "fake") <- assay(sce, "logcounts")
assay(sce, "fake")["Shh",] <- as.numeric(sce$gut_tube)

g = plotEmbryo(sce, 
               gene_name = "Shh", 
               assayName = "fake",
               quantileNorm = FALSE,
               embryo_subset = c("embryo1", "embryo2", "embryo3"),
               z_subset = c("2", "5"),
               outline = FALSE,
               colours = c("grey", "red")) + 
  add_scalebar(box.col = "black")
g
ggsave(g, file = "../Figures/gut_tube_selection.png", height = 10, width = 15)


# compare refined and normal cell type classes
tab = unclass(table(sce$celltype_mapped_refined, sce$celltype_mapped))
nms = intersect(rownames(tab), colnames(tab))

hc = hclust(dist(t(log10(1+tab))))
nms2 = intersect(hc$labels[hc$order],nms)

tab2 <- tab[c(nms2, setdiff(rownames(tab), nms)), c(nms2, setdiff(colnames(tab), nms))]

h = Heatmap(log10(1+t(tab2)),
            col = c("white", "grey", "black"),
            cluster_rows = FALSE,
            cluster_columns = FALSE,
            name = "Number of\ncells (log10)",
            row_title = "Automated cell classification",
            column_title = "Refined cell classification")

h

pdf("../Figures/refined_celltypes_comparison_heatmap.pdf", height = 8, width = 8.5)
print(h)
dev.off()

# compare refined and normal cell type classes
tab = unclass(table(sce$celltype_mapped_refined, sce$cluster_sub))

hc1 = hclust(dist(log10(1+tab)))
hc1_cut = cutree(hc1, 10)
hc1_cut_members = hc1_cut[as.character(sce$celltype_mapped_refined)]

hc2 = hclust(dist(t(log10(1+tab))))
hc2_cut = cutree(hc2, 10)
hc2_cut_members = hc2_cut[as.character(sce$cluster_sub)]

tab2 = unclass(table(hc1_cut_members, hc2_cut_members))
tab2_ordered = diagonalise(tab2)
tab2_ordered

rows = unlist(sapply(rownames(tab2_ordered), function(x) names(which(hc1_cut == x))))
cols = unlist(sapply(colnames(tab2_ordered), function(x) names(which(hc2_cut == x))))

tab3 = tab[rows,cols]

h = Heatmap(log10(1+(tab)),
            col = c("white", "grey", "black"),
            cluster_rows = TRUE,
            cluster_columns = TRUE,
            row_split = 9,
            column_split = substring(colnames(tab),1,1),
            name = "Number of\ncells (log10)",
            column_title = "Independent cell subclusters",
            row_title = "Refined cell classification")

h

pdf("../Figures/refined_celltypes_subcluster_comparison_heatmap.pdf", height = 6, width = 12)
print(h)
dev.off()

#######################
# plot UMAP of clusters with colours

UMAP_df = data.frame(
  UMAP1 = reducedDim(sce, "UMAP")[,1],
  UMAP2 = reducedDim(sce, "UMAP")[,2],
  cluster = sce$cluster,
  cluster_sub = sce$cluster_sub
)

UMAP_df_subcluster = data.frame(
  cluster_sub = sort(unique(sce$cluster_sub))
)
UMAP_df_subcluster$UMAP1 <- tapply(UMAP_df$UMAP1, UMAP_df$cluster_sub, mean)[UMAP_df_subcluster$cluster_sub]
UMAP_df_subcluster$UMAP2 <- tapply(UMAP_df$UMAP2, UMAP_df$cluster_sub, mean)[UMAP_df_subcluster$cluster_sub]
UMAP_df_subcluster$cluster <- unlist(lapply(strsplit(as.character(UMAP_df_subcluster$cluster_sub), "\\."), "[[", 1))

cluster_colours = colorRampPalette(RColorBrewer::brewer.pal(9, "Dark2"))(nrow(UMAP_df_subcluster))
names(cluster_colours) <- mixedsort(as.character(UMAP_df_subcluster$cluster_sub))

g = ggplot(UMAP_df[sample(rownames(UMAP_df)),], aes(x = UMAP1, y = UMAP2, colour = cluster_sub)) + 
  geom_point(size = 0.5, alpha = 0.3) + 
  theme_classic() + 
  theme(legend.position = "none") + 
  theme(axis.text = element_blank()) + 
  theme(axis.ticks = element_blank()) +
  scale_colour_manual(values = cluster_colours,
                      aesthetics = c("fill", "colour")) +
  coord_fixed() +
  guides(colour = guide_legend(title = "",
                               override.aes = list(size = 8, alpha = 1))) +
  geom_label_repel(colour = "black", data = UMAP_df_subcluster, aes(label = cluster_sub, fill = cluster_sub), alpha = 0.6) +
  NULL
g
ggsave(g, file = "../Figures/umap_subclusters.pdf",
       height = 6, width = 6)

# make a 3x3 of each cluster with different colours for the subtypes

gList = sapply(unique(UMAP_df$cluster), function(x){
  
  UMAP_df_2 = UMAP_df
  UMAP_df_2[UMAP_df_2$cluster != x, "cluster_sub"] <- NA
  
  g = ggplot(UMAP_df_2[sample(rownames(UMAP_df_2)),], aes(x = UMAP1, y = UMAP2, colour = cluster_sub)) + 
    geom_point(size = 0.5, alpha = 0.3) + 
    theme_classic() + 
    theme(legend.position = "none") + 
    theme(axis.text = element_blank()) + 
    theme(axis.ticks = element_blank()) +
    scale_colour_brewer(palette = "Set1", na.value = "lightgrey",
                        aesthetics = c("fill", "colour")) + 
    coord_fixed() +
    guides(colour = guide_legend(title = "",
                                 override.aes = list(size = 8, alpha = 1))) +
    geom_label_repel(colour = "black", data = subset(UMAP_df_subcluster, cluster == x), aes(label = cluster_sub, fill = cluster_sub), alpha = 0.6) +
    NULL
  g
}, simplify = FALSE)

g = wrap_plots(gList, nrow = 3)
ggsave(g, file = "../Figures/umap_subclusters_split.pdf",
       height = 10, width = 10)

# spatial plots of subclusters
boundary_polygons = getSegmentationVerticesDF(colData(sce),
                                              xname = "segmentation_vertices_x_global_affine",
                                              yname = "segmentation_vertices_y_global_affine",
                                              othercols = c("uniqueID","z","embryo", "cluster_sub"))

embryovals = unique(boundary_polygons$embryo)
zvals = unique(boundary_polygons$z)

for (embryoval in embryovals) {
  for (zval in zvals) {
    
    g = ggplot(subset(boundary_polygons, embryo == embryoval & z == zval)) + 
      geom_polygon(aes(group = uniqueID,
                       x = segmentation_vertices_x_global_affine,
                       y = -segmentation_vertices_y_global_affine,
                       fill = cluster_sub), colour = "black",
                   size = 0.1) + 
      scale_fill_manual(values = cluster_colours, na.value = "lightgrey") +
      theme_classic() +
      coord_fixed() +
      theme(legend.position = "none") +
      NULL
    
    gg = g + theme(axis.line = element_blank()) + 
      theme(axis.text = element_blank()) +
      theme(axis.ticks = element_blank()) + 
      theme(axis.title = element_blank()) +
      add_scalebar() +
      NULL
    gg
    ggsave(gg, file = paste0("../Figures/cluster_sub_coloured_",
                             embryoval, "_", zval, ".png"),
           height = 10, width = 8 )
    
  }  
}


ggg = as_ggplot(get_legend(gg + theme(legend.position = "right") + 
                             guides(fill = guide_legend(title = "", ncol = 4))))
print(ggg)

ggsave(ggg,  file = "../Figures/cluster_sub_coloured_legend.png",
       height = 4, width = 3)


marker_array = getMarkerArray(sce, 
                              group = "cluster_sub",
                              block = "embryo",
                              pseudobulk = TRUE)

markerArrayPlot(marker_array,
                grouplevel = "1.1",
                otherGroupLabel = "All other cells",
                onlyLabelTopRanked = TRUE,
                FDR = 0.1,
                LFC = 0.1)

# select significant marker genes for each subcluster
# top 3 positive genes 
topgenes = c()

for (i in dimnames(marker_array)[[2]]) {
  print(i)
  x = marker_array[,i,]
  if (sum(x[,"FDR"] < 0.1) == 0) {
    xx = x
  } else {
    xx = x[x[,"FDR"] < 0.1,, drop = FALSE]
  }
  topn = rownames(xx)[order(xx[,"LFC"], decreasing = TRUE)][1:3]
  topgenes <- sort(union(topgenes,topn))
  print(length(topgenes))
}



propExpressed = apply(logcounts(sce)[topgenes,],1,function(x){
  tapply(x, interaction(sce$cluster_sub, sce$embryo), function(x) mean(x))
})
propExpressed <- apply(propExpressed, 2, function(x) x/max(x))

cluster_colours_exp = cluster_colours[gsub("\\.embryo[0-9]", "", rownames(propExpressed))]
names(cluster_colours_exp) <- rownames(propExpressed)

h = Heatmap(t(propExpressed),
            col = c("gray75", "cornflowerblue", "black"),
            column_dend_height = unit(30, "mm"),
            column_split = substring(rownames(propExpressed),1,1),
            top_annotation = HeatmapAnnotation(
              which = "column",
              cluster_sub = rownames(propExpressed),
              col = list(cluster_sub = cluster_colours_exp),
              annotation_legend_param = list(cluster_sub = list(title = "")),
              show_legend = FALSE
            ),
            heatmap_legend_param = list(title = "Relative\nmean\nexpression",
                                        hjust = 0.5),
            column_title = "",
            show_heatmap_legend = TRUE,
            use_raster = TRUE,
            row_title_gp = gpar(fontsize = 10)
)

h

pdf("../Figures/cluster_sub_heatmap_top3.pdf",
    height = 12, width = 20)
print(h)
dev.off()

```

# Plot of the gut tube with the fitted principal curves

```{r}
DV_class = readRDS("../analysis_output/E8.5/gut_tube_DV_split.Rds")

D_cells = DV_class[["DV_class"]]

sce$gut_tube = ifelse(DV_class[["DV_class"]] == "Dorsal", 0, 1)
sce$gut_tube[is.na(sce$gut_tube)] <- 0.5
assay(sce, "fake") <- assay(sce, "logcounts")
assay(sce, "fake")["Shh",] <- as.numeric(sce$gut_tube)

g = plotEmbryo(sce, 
               gene_name = "Shh", 
               assayName = "fake",
               quantileNorm = FALSE,
               embryo_subset = c("embryo1", "embryo2", "embryo3"),
               z_subset = c("2"),
               outline = FALSE,
               colours = c("orange","grey80", "purple")) + 
  add_scalebar(box.col = "black")

fitList = readRDS("../analysis_output/E8.5/gut_tube_DV_princurve.Rds")

fit_df = do.call(rbind,sapply(1:length(fitList), function(i) {
  df = data.frame(x = fitList[[i]]$s[fitList[[i]]$ord,1],
                  y = fitList[[i]]$s[fitList[[i]]$ord,2],
                  embryo = sce[,rownames(fitList[[i]]$s)[1]]$embryo,
                  group = i,
                  type = DV_class[["DV_class"]][rownames(fitList[[i]]$s)[1]])
}, simplify = FALSE))


gg = g + geom_path(aes(x = x, y = y, group = type), data = fit_df, inherit.aes = FALSE,
) + 
  NULL

ggsave(gg, file = "../Figures/gut_tube_DV_AP_selection.png", height = 10, width = 15)
```


```{r}
# spatial graph of imputed genes
genes = c("Nkx2-1", "Pyy", "Rhox5", "Hhex", "Foxa3", "Pitx2", "Apoe", "Emb", "Apela")
for (gene in genes) {
  
  print(gene)
  
  fakegene = rownames(sce)[1]
  
  assay(sce, "mask") <- assay(sce, "logcounts")
  assay(sce, "mask")[fakegene, ] <- imp[gene,colnames(sce)]
  assay(sce, "mask")[fakegene, sce$celltype_mapped_refined != "Gut tube"] <- NA
  
  g = plotEmbryo(sce, 
                 gene_name = fakegene, 
                 assayName = "mask",
                 quantileNorm = FALSE,
                 embryo_subset = "embryo1",
                 z_subset = "2",
                 outline = TRUE)
  g
  ggsave(g, file = paste0("../Figures/corrected_gut_tube_", gene, "_embryo1_z2.pdf"),
         height = 20, width = 24)
  
}

genes = c("Nkx2-1", "Nkx2-5", "Nog")
genes %in% rownames(sce)
genes %in% rnames
for (gene in genes) {
  print(gene)
  
  fakegene = rownames(sce)[1]
  
  assay(sce, "mask") <- assay(sce, "logcounts")
  assay(sce, "mask")[fakegene, ] <- imp[gene,colnames(sce)]
  
  g = plotEmbryo(sce, 
                 gene_name = fakegene, 
                 assayName = "mask",
                 quantileNorm = FALSE,
                 embryo_subset = "embryo1",
                 z_subset = "2",
                 outline = TRUE)
  g
  ggsave(g, file = paste0("../Figures/imputed_", gene, "_embryo1_z2.pdf"),
         height = 20, width = 24)
  
}
```

# plot of brain subclusters

```{r}
brain_sub_names_order = c(
  "Rhombencephalon 1",
  "Rhombencephalon 2",
  "Rhombencephalon 3",
  "Tegmentum",
  "Mesencephalon",
  "Prosencephalon 1",
  "Prosencephalon 2"
)

mapped.brain_named = readRDS("../analysis_output/E8.5/E8.5_mapped.brain_named.Rds")

sce$mapped.brain_named = mapped.brain_named[colnames(sce)]

brain_sub_names_colours = nord::nord_palettes["algoma_forest"][[1]][1:length(brain_sub_names_order)]
names(brain_sub_names_colours) <- brain_sub_names_order

saveRDS(brain_sub_names_colours, file = "../analysis_output/E8.5/E8.5_brain_subcluster_colours.Rds")

g = ggplot(data.frame(x = factor(names(brain_sub_names_colours),
                                 levels = names(brain_sub_names_colours))), aes(x = x, fill = x)) +
  geom_bar(y = 1) + scale_fill_manual(values = brain_sub_names_colours) + 
  guides(fill = guide_legend(title = ""))

gg = as_ggplot(get_legend(g + theme(legend.position = "right") + 
                            guides(fill = guide_legend(title = ""))))
print(gg)

ggsave(gg,  file = "../Figures/brain_coloured_legend.png",
       height = 4, width = 3)


celltype_mapped_col <- brain_sub_names_colours[as.character(sce$mapped.brain_named)]
names(celltype_mapped_col) <- colnames(sce)

boundary_polygons = getSegmentationVerticesDF(colData(sce),
                                              xname = "segmentation_vertices_x_global_affine",
                                              yname = "segmentation_vertices_y_global_affine",
                                              othercols = c("uniqueID","z","embryo"))

embryovals = unique(boundary_polygons$embryo)
zvals = unique(boundary_polygons$z)


for (embryoval in embryovals) {
  for (zval in zvals) {
    
    g = ggplot(subset(boundary_polygons, embryo == embryoval & z == zval)) + 
      geom_polygon(aes(group = uniqueID,
                       x = segmentation_vertices_x_global_affine,
                       y = -segmentation_vertices_y_global_affine,
                       fill = uniqueID), colour = "black",
                   size = 0.1) + 
      scale_fill_manual(values = celltype_mapped_col, na.value = "lightgrey") +
      theme_classic() +
      coord_fixed() +
      theme(legend.position = "none") +
      NULL
    
    gg = g + theme(axis.line = element_blank()) + 
      theme(axis.text = element_blank()) +
      theme(axis.ticks = element_blank()) + 
      theme(axis.title = element_blank()) +
      add_scalebar() +
      NULL
    gg
    ggsave(gg, file = paste0("../Figures/mapped.brain_coloured_",
                             embryoval, "_", zval, ".png"),
           height = 10, width = 8 )
    
    if (embryoval == "embryo2" & zval == 2) {
      ggg = gg + geom_polygon(aes(x = x, y = y), inherit.aes = FALSE, 
                              data = data.frame(x = c(-0.5, -0.5, 1.15, 1.15),
                                                y = c(2,3.15, 3.15, 2)),
                              fill = NA,
                              colour = "red",
                              size = 1.5)
      
      ggsave(ggg, file = paste0("../Figures/mapped.brain_coloured_box_",
                                embryoval, "_", zval, ".png"),
             height = 10, width = 8 )
      
    } 
    
    x_range = c(min(g$data$segmentation_vertices_x_global_affine), 1)
    y_range = c(0, max(g$data$segmentation_vertices_y_global_affine))
    
    
    gList = sapply(names(brain_sub_names_colours), function(nm) {
      
      celltype_mapped_col_sub = celltype_mapped_col
      celltype_mapped_col_sub[celltype_mapped_col_sub != brain_sub_names_colours[nm]] <- NA
      
      celltype_mapped_col_sub_black <- celltype_mapped_col_sub
      celltype_mapped_col_sub_black[!is.na(celltype_mapped_col_sub_black)] <- "black"
      
      g = ggplot(subset(boundary_polygons, embryo == embryoval & z == zval)) + 
        geom_polygon(aes(group = uniqueID,
                         x = segmentation_vertices_x_global_affine,
                         y = -segmentation_vertices_y_global_affine,
                         fill = uniqueID,
                         colour = uniqueID), #colour = "lightgrey",
                     size = 0.1) + 
        scale_fill_manual(values = celltype_mapped_col_sub, na.value = "lightgrey") +
        scale_colour_manual(values = celltype_mapped_col_sub_black, na.value = "lightgrey") +
        theme_classic() +
        coord_fixed() +
        theme(legend.position = "none") +
        ggtitle(nm) +
        theme(plot.title = element_text(hjust = 0.5)) +
        NULL
      
      
      gg = g + theme(axis.line = element_blank()) + 
        theme(axis.text = element_blank()) +
        theme(axis.ticks = element_blank()) + 
        theme(axis.title = element_blank()) +
        scale_x_continuous(limits = x_range, breaks = x_range) +
        scale_y_continuous(limits = y_range, breaks = y_range) +
        add_scalebar(x = x_range[1] + 0.4*diff(x_range), y = y_range[2] - 0.5) +
        coord_fixed() +
        theme(plot.margin = unit(c(1,0,0,0), "cm")) +
        NULL
      gg
      
      return(gg)
    },simplify = FALSE)
    
    glong = wrap_plots(gList, nrow = 1)
    
    ggsave(glong, file = paste0("../Figures/brain_coloured_split_",
                                embryoval, "_", zval, ".png"),
           height = 4, width = 16)
    ggsave(glong, file = paste0("../Figures/brain_coloured_split_",
                                embryoval, "_", zval, ".pdf"),
           height = 4, width = 16)
  }
}

discriminating_array_list = list()

groupValue = "mapped.brain_named"

groups = sort(unique(colData(sce)[, groupValue]))

for (grouplevel in groups) {
  
  print(grouplevel)
  
  groups2 = setdiff(groups, grouplevel)
  
  for (grouplevel2 in groups2) {
    
    print(grouplevel2)
    
    ma = getMarkerArray(sce,
                        group = groupValue,
                        block = "embryo",
                        subset = colData(sce)[, groupValue] %in% c(grouplevel,
                                                                   grouplevel2),
                        pseudobulk = TRUE,
                        verbose = FALSE)
    
    discriminating_array_list[[grouplevel]][[grouplevel2]] <- ma
    
  }
  
}


marker_array = getMarkerArray(sce, 
                              group = "mapped.brain_named",
                              block = "embryo",
                              subset = !is.na(sce$mapped.brain_named),
                              pseudobulk = TRUE)

markerArrayPlot(marker_array,
                grouplevel = "Tegmentum",
                otherGroupLabel = "All other cells",
                onlyLabelTopRanked = TRUE,
                FDR = 0.1,
                LFC = 0.1)

# select significant marker genes for each subcluster
# top 3 positive genes 
topgenes = c()

for (j in names(discriminating_array_list)) {
  for (i in names(discriminating_array_list[[j]])) {
    print(i)
    x = discriminating_array_list[[j]][[i]][,j,]
    if (sum(x[,"FDR"] < 0.1) == 0) {
      xx = x
    } else {
      xx = x[x[,"FDR"] < 0.1,, drop = FALSE]
    }
    topn = rownames(xx)[order(xx[,"LFC"], decreasing = TRUE)][1:5]
    topgenes <- sort(union(topgenes,topn))
    print(length(topgenes))
  }
}



propExpressed = apply(logcounts(sce)[topgenes,],1,function(x){
  tapply(x, interaction(sce$mapped.brain_named, sce$embryo), function(x) mean(x))
})
propExpressed <- apply(propExpressed, 2, function(x) x/max(x))

cluster_colours_exp = brain_sub_names_colours[gsub("\\.embryo[0-9]", "", rownames(propExpressed))]
names(cluster_colours_exp) <- rownames(propExpressed)

h = Heatmap(t(propExpressed),
            col = c("gray75", "cornflowerblue", "black"),
            column_dend_height = unit(30, "mm"),
            top_annotation = HeatmapAnnotation(
              which = "column",
              cluster_sub = rownames(propExpressed),
              col = list(cluster_sub = cluster_colours_exp),
              annotation_legend_param = list(cluster_sub = list(title = "")),
              show_legend = FALSE
            ),
            heatmap_legend_param = list(title = "Relative\nmean\nexpression",
                                        hjust = 0.5),
            column_title = "",
            show_heatmap_legend = TRUE,
            use_raster = TRUE,
            row_title_gp = gpar(fontsize = 10)
)

h

pdf("../Figures/brain_mapped_heatmap_top.pdf",
    height = 12, width = 10)
print(h)
dev.off()
```

# Celltype proportions

```{r}
ctypes_all = sort(unique(atlas_sub$celltype_parsed))
celltype_props = data.frame(
  celltype = ctypes_all,
  atlas_E8.5 = unclass(table(factor(atlas_sub$celltype_parsed, levels = ctypes_all))),
  spatial_all = unclass(table(factor(sce$celltype_mapped_refined, levels = ctypes_all))),
  spatial_embryo1 = unclass(table(factor(sce$celltype_mapped_refined[sce$embryo == "embryo1"], levels = ctypes_all))),
  spatial_embryo2 = unclass(table(factor(sce$celltype_mapped_refined[sce$embryo == "embryo2"], levels = ctypes_all))),
  spatial_embryo3 = unclass(table(factor(sce$celltype_mapped_refined[sce$embryo == "embryo3"], levels = ctypes_all))
  ))

plot(1+celltype_props$atlas_E8.5, 1+celltype_props$spatial_all, log = "xy")

g1 = ggplot(celltype_props, aes(x = 1 + atlas_E8.5, y = 1 + spatial_embryo1)) + 
  geom_point() + 
  geom_text_repel(aes(label = celltype)) + 
  theme_classic() +
  scale_x_log10() + 
  scale_y_log10() +
  geom_smooth(method = "lm") +
  xlab("Number of cells in E8.5 atlas") +
  ylab("Number of cells in E8.5 embryo 1") +
  NULL

g2 = ggplot(celltype_props, aes(x = 1 + atlas_E8.5, y = 1 + spatial_embryo2)) + 
  geom_point() + 
  geom_text_repel(aes(label = celltype)) + 
  theme_classic() +
  scale_x_log10() + 
  scale_y_log10() +
  geom_smooth(method = "lm") +
  xlab("Number of cells in E8.5 atlas") +
  ylab("Number of cells in E8.5 embryo 2") +
  NULL

g3 = ggplot(celltype_props, aes(x = 1 + atlas_E8.5, y = 1 + spatial_embryo3)) + 
  geom_point() + 
  geom_text_repel(aes(label = celltype)) + 
  theme_classic() +
  scale_x_log10() + 
  scale_y_log10() +
  geom_smooth(method = "lm") +
  xlab("Number of cells in E8.5 atlas") +
  ylab("Number of cells in E8.5 embryo 3") +
  NULL

g1 + g2 + g3

g = ggplot(celltype_props, aes(x = 1 + atlas_E8.5)) +
  geom_point(aes(y = 1 + spatial_embryo1), colour = "orange") + 
  geom_point(aes(y = 1 + spatial_embryo2), colour = "red") + 
  geom_point(aes(y = 1 + spatial_embryo3), colour = "blue") + 
  theme_classic() +
  scale_x_log10() + 
  scale_y_log10() +
  geom_smooth(method = "lm", aes(y = 1 + spatial_embryo1)) +
  geom_smooth(method = "lm", aes(y = 1 + spatial_embryo2)) +
  geom_smooth(method = "lm", aes(y = 1 + spatial_embryo3)) +
  xlab("Number of cells in E8.5 atlas") +
  ylab("Number of cells in E8.5 embryo 3") +
  NULL
g

lm1 = lm(I(log10(1+spatial_embryo1)) ~ I(log10(1+atlas_E8.5)), data = celltype_props)
celltype_props$lm1_res <- lm1$residuals

lm2 = lm(I(log10(1+spatial_embryo2)) ~ I(log10(1+atlas_E8.5)), data = celltype_props)
celltype_props$lm2_res <- lm2$residuals

lm3 = lm(I(log10(1+spatial_embryo3)) ~ I(log10(1+atlas_E8.5)), data = celltype_props)
celltype_props$lm3_res <- lm3$residuals

lm_df = data.frame(
  celltype = c(names(lm1$residuals),names(lm2$residuals),names(lm3$residuals)),
  residual = c(lm1$residuals,lm2$residuals,lm3$residuals),
  embryo = rep(c("embryo1","embryo2","embryo3"), 
               times = c(length(lm1$residuals),length(lm2$residuals),length(lm3$residuals)))
)

lm_df$celltype <- factor(lm_df$celltype, levels = rev(c(names(sort(tapply(lm_df$residual, lm_df$celltype, mean))))))

g = ggplot(lm_df, aes(x = celltype, group = embryo, fill = celltype, y = residual)) + 
  geom_hline(yintercept = c(-1,1), linetype = "dashed", colour = "grey") +
  geom_col(position = "dodge", colour = "black") + 
  theme_classic() + 
  theme(axis.text.x = element_text(angle = 60, hjust = 1, size = 10)) +
  xlab("") +
  theme(legend.position = "none") +
  scale_alpha_manual(values = c("TRUE" = 1, "FALSE" = 0.3)) +
  scale_fill_manual(values = celltype_colours) +
  ylab("Relative enrichment") +
  NULL
g

ggsave(g, file = "../Figures/celltype_enrichment.pdf", height = 4, width = 10)
```

# Density plot of cell types along axis

```{r}
# AP position info
AP_info = readRDS("../analysis_output/E8.5/gut_tube_DV_split.Rds")

# gut tube subtypes
colData(sce)$gut_tube_subtype

df = cbind(as.data.frame(colData(sce)),
           DV_class = AP_info[["DV_class"]][colnames(sce)],
           AP_value = AP_info[["AP_value"]][colnames(sce)])

df$AP_value_norm = unsplit(tapply(df$AP_value, list(df$embryo, df$DV_class), function(x) x - min(x)),
                           list(df$embryo, df$DV_class))
df$gut_tube_subtype <- factor(df$gut_tube_subtype, levels = names(AP_pseudo_colours))

df$AP_value_rank = unsplit(tapply(df$AP_value_norm, list(df$embryo, df$DV_class), rank),
                           list(df$embryo, df$DV_class))

df <- df[!is.na(df$gut_tube_subtype) & !is.na(df$DV_class),]

df$gut_tube_subtype_conf = now_joint_df[rownames(df), "confany"]

ggplot(df, aes(x = gut_tube_subtype, y = AP_value_norm, fill = DV_class)) + 
  geom_violin() + 
  facet_grid(. ~ embryo, scales = "free") + 
  NULL

ggplot(df, aes(x = AP_value_norm)) + 
  geom_density(aes(fill = gut_tube_subtype, colour = gut_tube_subtype), alpha = 0.5) + 
  scale_fill_manual(values = AP_pseudo_colours, aesthetics = c("fill", "colour")) +
  facet_grid(embryo ~ DV_class) +
  theme_classic() +
  coord_flip() +
  NULL

ggplot(df, aes(y = AP_value_rank, x = gut_tube_subtype)) + 
  geom_tile(fill = "black", colour = "black") +
  facet_grid(~interaction(embryo, DV_class)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1)) + 
  NULL

ggplot(df, aes(y = AP_value_rank, x = gut_tube_subtype)) + 
  geom_tile(aes(fill = gut_tube_subtype_conf, colour = gut_tube_subtype_conf)) +
  facet_grid(~interaction(embryo, DV_class)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1)) + 
  NULL

df$DV_class_rev = factor(df$DV_class, levels = rev(levels(df$DV_class)))
df$gut_tube_subtype_rev = factor(df$gut_tube_subtype, levels = rev(levels(df$gut_tube_subtype)))

# calculate the proportion of dorsal for each embryo
df_prop = data.frame(gut_tube_subtype = rep(levels(df$gut_tube_subtype), each = 3),
                     embryo = rep(unique(df$embryo), times = length(levels(df$gut_tube_subtype))))
df_prop$prop = apply(as.matrix(df_prop),1,function(x){
  df_subset = subset(df, gut_tube_subtype == x[1] & embryo == x[2])
  return(sum(df_subset$DV_class == "Dorsal")/nrow(df_subset))
})
df_prop <- na.omit(df_prop)
df_prop$gut_tube_subtype_rev <- factor(df_prop$gut_tube_subtype, levels = rev(levels(df_prop$gut_tube_subtype)))

g = ggplot(df, aes(x = gut_tube_subtype_rev)) + 
  geom_bar(aes(fill = DV_class_rev), position = "fill", alpha = 0.8) + 
  geom_point(aes(y = prop), data = df_prop, size = 3, alpha = 0.5) +
  theme_classic() + 
  coord_flip() +
  xlab("") + 
  ylab("") + 
  scale_fill_manual(values = c("Dorsal" = "orange", "Ventral" = "purple")) +
  guides(fill = guide_legend(title = "")) +
  scale_y_continuous(limits = c(0,1), expand = c(0,0)) +
  theme(axis.line.y = element_blank()) +
  theme(legend.position = "right") +
  theme(plot.margin = unit(c(0,0.5,0,0), "cm")) + 
  NULL
g
ggsave(g, file = "../Figures/DV_class_gut_barplot.pdf", height = 4, width = 8)


for (embryoval in c("embryo1", "embryo2", "embryo3")) {
  
  df1 = subset(df, embryo == embryoval & DV_class == "Dorsal" & gut_tube_subtype != "Large intestine/Colon")
  df1$gut_tube_subtype <- factor(df1$gut_tube_subtype, levels = levels(droplevels(df$gut_tube_subtype)))
  
  g1 = ggplot(df1, aes(y = -AP_value_rank, x = gut_tube_subtype)) + 
    geom_tile(aes(fill = gut_tube_subtype_conf, colour = gut_tube_subtype_conf)) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 60, hjust = 1, size = 15)) + 
    theme(legend.position = "none") +
    scale_colour_gradient(high = "red", low = "black", aesthetics = c("fill", "colour")) + 
    scale_x_discrete(drop=FALSE) +
    theme(axis.text.y = element_blank()) +
    theme(plot.title = element_text(size = 20)) +
    theme(axis.title.y  = element_text(size = 20)) +
    theme(axis.ticks = element_blank()) +
    ylab("AP cell ranking") +
    xlab("") +
    ggtitle("Dorsal") +
    theme(plot.title = element_text(hjust = 0.5)) +
    scale_y_continuous(limits = rev(c(0,-max(df1$AP_value_rank))), expand = c(0, 0)) +
    NULL
  
  
  df2 = subset(df, embryo == embryoval & DV_class == "Ventral" & gut_tube_subtype != "Large intestine/Colon")
  df2$gut_tube_subtype <- factor(df2$gut_tube_subtype, levels = levels(droplevels(df$gut_tube_subtype)))
  g2 = ggplot(df2,
              aes(y = -AP_value_rank, x = gut_tube_subtype)) + 
    geom_tile(aes(fill = gut_tube_subtype_conf, colour = gut_tube_subtype_conf)) +
    theme_classic() + 
    theme(axis.text.x = element_text(angle = 60, hjust = 1, size = 15)) + 
    scale_colour_gradient(high = "red", low = "black", aesthetics = c("fill", "colour")) + 
    scale_x_discrete(drop=FALSE) +
    theme(axis.text.y = element_blank()) +
    theme(axis.ticks = element_blank()) +
    ylab("") +
    theme(plot.title = element_text(size = 20)) +
    ggtitle("Ventral") +
    theme(plot.title = element_text(hjust = 0.5)) +
    xlab("") +
    guides(fill = guide_colourbar(title = "", direction = "vertical", reverse = FALSE),
           colour = guide_colourbar(title = "", direction = "vertical", reverse = FALSE)) +
    scale_y_continuous(limits = rev(c(0,-max(df2$AP_value_rank))), expand = c(0, 0)) +
    NULL
  
  g = g1 + g2
  g
  
  ggsave(g, file = paste0("../Figures/AP_pseudo_tile_positions_", embryoval, ".pdf"), height = 8, width = 5)
}
```

```{r}
# logistic regression error rate for class along AP axis
# perform along the AP-axis pairs
logit = function(p) log(p/(1 - p)) 
invlogit = function(x) 1/(1 + exp(-x)) 

out = NULL

for (embryoval in c("embryo1", "embryo2", "embryo3")) {
  for (DV_class in c("Dorsal", "Ventral")) {
    for (type1 in levels(df$gut_tube_subtype)[1:(length(levels(df$gut_tube_subtype))- 1)]) {
      
      type2 = levels(df$gut_tube_subtype)[which(levels(df$gut_tube_subtype) == type1) + 1]
      
      df1 = subset(df, embryo == embryoval & DV_class == "Dorsal")
      df1$gut_tube_subtype <- factor(df1$gut_tube_subtype, levels = levels(df$gut_tube_subtype))
      df2 = subset(df, embryo == embryoval & DV_class == "Ventral")
      df2$gut_tube_subtype <- factor(df2$gut_tube_subtype, levels = levels(df$gut_tube_subtype))
      
      if (DV_class == "Dorsal") {
        df_sub = subset(df1, gut_tube_subtype %in% c(type1, type2))
      } else {
        df_sub = subset(df2, gut_tube_subtype %in% c(type1, type2))
      }
      
      
      df_sub$gut_tube_subtype <- factor(df_sub$gut_tube_subtype, levels = c(type1, type2))
      
      if (length(unique(df_sub$gut_tube_subtype)) != 2) next
      
      fit = glm(gut_tube_subtype ~ AP_value_norm, data = df_sub, family = "binomial")
      
      predvals = invlogit(predict(fit, df_sub)) < 0.5
      table(predvals, df_sub$gut_tube_subtype)
      err = mean(predvals != (df_sub$gut_tube_subtype == type1))
      
      err
      
      out <- rbind(out, c(type1 = type1, type2 = type2, err = err, DV_class = DV_class,
                          embryo = embryoval))
    }
  }
}

out_summary = data.frame(
  type1 = out[, "type1"],
  pair = paste0(out[, "type1"], ", ", out[, "type2"]),
  err = as.numeric(out[, "err"]),
  DV_class = out[, "DV_class"],
  embryo = out[, "embryo"]
)
out_summary$pair <- factor(out_summary$pair, 
                           levels = paste0(levels(df$gut_tube_subtype)[1:8],
                                           ", ",
                                           levels(df$gut_tube_subtype)[2:9]))

g = ggplot(out_summary, aes(x = pair, y = err)) +
  geom_point(aes(colour = embryo), size = 5, alpha = 0.6) + 
  facet_wrap(~DV_class) + 
  theme_minimal() + 
  theme(axis.line = element_line()) +
  theme(axis.text.x = element_text(angle = 60, hjust = 1)) + 
  xlab("") + 
  ylab("") + 
  NULL
g

ggsave(g, file = "../Figures/AP_logistic_error_pairs.pdf", height = 6, width = 8)
```

# Subclustering

```{r}
colour_scheme = c("gray75","cornflowerblue","black")

tab = unclass(table(sce$cluster, sce$celltype_mapped))
tab

Heatmap(log10(1+tab), cluster_rows = TRUE, cluster_columns = TRUE,
        col = colour_scheme,
        heatmap_legend_param = list(title = "Number of cells (log10)")
)

tab_sub = unclass(table(sce$cluster_sub, sce$celltype_mapped))
tab_sub

h = Heatmap(log2(1+tab_sub),
            split = substring(sort(unique(sce$cluster_sub)),1,1),
            col = colour_scheme,
            heatmap_legend_param = list(title = "Number of cells (log10)"))

print(h)

pdf("../Figures/subclustering_celltypes_heatmap.pdf", height = 10, width = 16)
print(h)
dev.off()

h = Heatmap(log2(1+tab_sub),
            split = substring(sort(unique(sce$cluster_sub)),1,1),
            col = colour_scheme,
            heatmap_legend_param = list(title = "Number of cells (log10)",
                                        direction = "horizontal"),
            row_names_side = "left",
            row_dend_side = "right",
            column_names_side = "top",
            column_dend_side = "bottom",
            column_names_rot = 45)

draw(h, heatmap_legend_side = "bottom",
     padding = unit(c(2, 2, 2, 20), "mm"))

pdf("../Figures/subclustering_celltypes_heatmap_swap.pdf", height = 8, width = 12)
draw(h, heatmap_legend_side = "bottom",
     padding = unit(c(2, 2, 2, 20), "mm"))
dev.off()
```

# mRNA digital in situ for other genes

```{r}
genes_colours_all = c("red", "cyan", "orange","green")

allsinglegenes = c("Emcn", "Hand1", "Msx1", "Postn", "Cdh5", "Sfrp5", "Dlk1", "Kdr", "Gata4", "Hoxb1", "Pecam1", "Gata5", "Snai1", "Sox17", "Hcn4", "Tbx5", "Ttn", "Nkx2-5")

allgenes = list(
  c("Tbx5", "Cdh5", "Postn", "Dlk1"),
  split(allsinglegenes, allsinglegenes)
)


for (genes in allgenes[1]) {
  
  genes = intersect(genes, rownames(sce))
  
  print(genes)
  
  genes_colours = genes_colours_all[1:length(genes)]
  
  names(genes_colours) <- genes
  
  heart_shift_x = c(0,-1.2, 0)
  names(heart_shift_x) <- c("embryo1","embryo2","embryo3")
  
  heart_shift_y = c(0,-0.25, -0.6)
  names(heart_shift_y) <- c("embryo1","embryo2","embryo3")
  
  range_x = c(-1.9, 0.5)
  range_y = c(-1.5, 0.95)
  
  mRNA_genes = mRNA_df[mRNA_df$geneID %in% genes,]
  boundary_polygons_sub = boundary_polygons
  
  mRNA_genes$x_global_affine <- mRNA_genes$x_global_affine + heart_shift_x[mRNA_genes$embryo]
  mRNA_genes$y_global_affine <- mRNA_genes$y_global_affine + heart_shift_y[mRNA_genes$embryo]
  
  boundary_polygons_sub$segmentation_vertices_x_global_affine <- boundary_polygons_sub$segmentation_vertices_x_global_affine + heart_shift_x[boundary_polygons_sub$embryo]
  boundary_polygons_sub$segmentation_vertices_y_global_affine <- boundary_polygons_sub$segmentation_vertices_y_global_affine + heart_shift_y[boundary_polygons_sub$embryo]
  
  g = ggplot(boundary_polygons_sub,
             aes(x = segmentation_vertices_x_global_affine,
                 y = -segmentation_vertices_y_global_affine,
                 group = uniqueID),
             fill = "lightgrey",
             size = 0.1) + 
    geom_polygon() + 
    
    geom_polygon(data = subset(boundary_polygons_sub, celltype_mapped == "Cardiomyocytes"),
                 colour = "lightblue",
                 size = 0.1) +
    
    theme_classic() +
    facet_grid(z~embryo) +
    guides(fill = guide_legend(title = "Ttn normalised logcounts")) +
    coord_fixed() +
    xlab("") +
    ylab("") +
    xlim(range_x) + 
    ylim(range_y) +
    theme(plot.background = element_rect(fill = "black"),
          panel.background = element_rect(fill = "black")) +
    NULL
  
  g_dots = g + 
    geom_point(aes(x = x_global_affine, y = -y_global_affine,
                   colour = geneID), 
               data = mRNA_genes,
               size = 0.01,
               # colour = "red",
               inherit.aes = FALSE) + 
    scale_colour_manual(values = genes_colours) +
    guides(colour = guide_legend(override.aes = list(size = 5))) +
    NULL
  
  print(g_dots)
  
  ggsave(g_dots, file = paste0("../Figures/cardio_insitu/", paste0(sort(genes), collapse = "_"),".png"), height = 8, width = 12)
}
```

# Cell type graphs

```{r}
for (embryo_name in unique(boundary_polygons$embryo)) {
  
  print(embryo_name)
  
  g = ggplot(data = subset(boundary_polygons, 
                           z == 2 & embryo == embryo_name)) + 
    geom_polygon(aes(x = segmentation_vertices_x_global_affine,
                     y = -segmentation_vertices_y_global_affine,
                     group = uniqueID,
                     fill = celltype_mapped),
                 colour = "black", size = 0.1) +
    theme_transparent() +
    scale_fill_manual(values = celltype_colours) +
    theme_classic() +
    coord_fixed() +
    theme(legend.position = "none") +
    theme(axis.line = element_blank()) + 
    theme(axis.text = element_blank()) + 
    theme(axis.title = element_blank()) +
    theme(axis.ticks = element_blank()) +
    NULL
  
  print(g)
  
  ggsave(g, file = paste0("../Figures/celltype_mapped_",
                          embryo_name,
                          ".png"),
         height = 10, width = 8)
}

# also for the gut tube types

gut_tube_celltypes = c("Gut",
                       "Pharyngeal mesoderm",
                       "Paraxial mesoderm",
                       "Def. endoderm",
                       "ExE endoderm",
                       "Parietal endoderm",
                       "Visceral endoderm")

boundary_polygons$celltype_mapped_gut_masked = boundary_polygons$celltype_mapped
boundary_polygons$celltype_mapped_gut_masked[!boundary_polygons$celltype_mapped_gut_masked %in% gut_tube_celltypes] <- NA

for (embryo_name in rev(unique(boundary_polygons$embryo))) {
  
  print(embryo_name)
  
  g = ggplot(data = subset(boundary_polygons, 
                           z == 2 & embryo == embryo_name)) + 
    geom_polygon(aes(x = segmentation_vertices_x_global_affine,
                     y = -segmentation_vertices_y_global_affine,
                     group = uniqueID,
                     fill = celltype_mapped_gut_masked),
                 colour = "black", size = 0.1) +
    theme_transparent() +
    scale_fill_manual(values = celltype_colours, na.value = "white") +
    theme_classic() +
    coord_fixed() +
    theme(legend.position = "none") +
    theme(axis.line = element_blank()) + 
    theme(axis.text = element_blank()) + 
    theme(axis.title = element_blank()) +
    theme(axis.ticks = element_blank()) +
    NULL
  
  print(g)
  
  ggsave(g, file = paste0("../Figures/celltype_mapped_gut_types_",
                          embryo_name,
                          ".png"),
         height = 10, width = 8)
}

gg = as_ggplot(get_legend(g + theme(legend.position = "right") + 
                            guides(fill = guide_legend(title = ""))))
print(gg)

ggsave(gg,  file = "../Figures/celltype_mapped_gut_types_legend.png",
       height = 4, width = 3)
```

# Network graph

```{r}
neighbours_raw = readRDS("../analysis_output/E8.5/E8.5_neighbourGraph_1.3.Rds")
neighbours_pairs = get.edgelist(neighbours_raw)

sce_sub = sce[,sce$embryo == "embryo1" & sce$z == 2]

neighbours_subset_idx = (neighbours_pairs[,1] %in% sce_sub$uniqueID) & 
  (neighbours_pairs[,2] %in% sce_sub$uniqueID)
table(neighbours_subset_idx)

neighbours <- neighbours_pairs[neighbours_subset_idx,]

neighbours_df = data.frame(
  x1 = colData(sce_sub)[neighbours[,1], "x_global_affine"],
  x2 = colData(sce_sub)[neighbours[,2], "x_global_affine"],
  y1 = colData(sce_sub)[neighbours[,1], "y_global_affine"],
  y2 = colData(sce_sub)[neighbours[,2], "y_global_affine"],
  ct1 = colData(sce_sub)[neighbours[,1], "celltype_mapped_refined"],
  ct2 = colData(sce_sub)[neighbours[,2], "celltype_mapped_refined"]
)

neighbours_df$ctSame = ifelse(as.character(neighbours_df$ct1) == as.character(neighbours_df$ct2),
                              "Same","Different")

g = ggplot(data = subset(boundary_polygons, 
                         z == 2 & embryo == "embryo1")) + 
  geom_polygon(aes(x = segmentation_vertices_x_global_affine,
                   y = -segmentation_vertices_y_global_affine,
                   group = uniqueID),
               colour = "black", size = 0.1, fill = "white") +
  theme_transparent() +
  theme_classic() +
  coord_fixed() +
  theme(legend.position = "none") +
  theme(axis.line = element_blank()) + 
  theme(axis.text = element_blank()) + 
  theme(axis.title = element_blank()) +
  theme(axis.ticks = element_blank()) +
  NULL

g2 = g + geom_segment(aes(x = x1, y = -y1, xend = x2, yend = -y2,
                          colour = ctSame), data = neighbours_df,
                      inherit.aes = FALSE,
                      size = 0.1) + 
  scale_colour_manual(values = c("Same" = "Blue", "Different" = "red")) +
  NULL

g2


g = ggplot(data = subset(boundary_polygons, 
                         z == 2 & embryo == "embryo1")) + 
  geom_polygon(aes(x = segmentation_vertices_x_global_affine,
                   y = -segmentation_vertices_y_global_affine,
                   group = uniqueID,
                   fill = uniqueID),
               colour = "black", size = 0.1) +
  theme_transparent() +
  scale_fill_manual(values = celltype_mapped_col) +
  theme_classic() +
  coord_fixed() +
  theme(legend.position = "none") +
  theme(axis.line = element_blank()) + 
  theme(axis.text = element_blank()) + 
  theme(axis.title = element_blank()) +
  theme(axis.ticks = element_blank()) +
  NULL

g2 = g + geom_segment(aes(x = x1, y = -y1, xend = x2, yend = -y2,
                          colour = ctSame), data = subset(neighbours_df, ctSame == "Different"),
                      inherit.aes = FALSE,
                      size = 0.5) + 
  scale_colour_manual(values = c("Same" = "Blue", "Different" = "black")) +
  NULL
g2


g2 = g + geom_segment(aes(x = x1, y = -y1, xend = x2, yend = -y2), data = neighbours_df,
                      inherit.aes = FALSE,
                      size = 0.1,
                      colour = "black") + 
  NULL
g2
ggsave(g2, file = "../Figures/neighbourhood_zoom.pdf", height = 20, width = 8)

g2 = g + geom_segment(aes(x = x1, y = -y1, xend = x2, yend = -y2,
                          size = ctSame), 
                      data = neighbours_df,
                      inherit.aes = FALSE) + 
  scale_colour_manual(values = c("Same" = "black", "Different" = "black")) +
  scale_size_manual(values = c("Same" = 0.3, "Different" = 0.3)) +
  xlim(c(-1.35, -0.15)) +
  ylim(c(-0.50, 0.50)) +
  NULL
g2
```



```{r}
ggplot(as.data.frame(colData(sce)), 
       aes(x = embryo, y = log10(1+total))) +
  geom_boxplot(outlier.size = 0.5) + 
  ggtitle("Total counts per cell") +
  theme_classic() +
  NULL

cData = as.data.frame(colData(sce))
cData$embryo_pos <- factor(cData$embryo_pos, levels = mixedsort(unique(cData$embryo_pos)))

g = ggplot(cData,
           aes(x = embryo_pos, y = total,
               fill = embryo)) +
  geom_boxplot(outlier.size = 0.5) + 
  ggtitle("Total counts per FOV") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_y_log10() +
  theme(legend.position = "bottom") +
  ylab("") +
  xlab("") +
  NULL
g
ggsave(g, file = "../Figures/total_counts.pdf", height = 6, width = 10)

g = ggplot(cData,
           aes(x = embryo_pos, y = detected,
               fill = embryo)) +
  geom_boxplot(outlier.size = 0.5) + 
  ggtitle("Total genes detected per FOV") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_y_log10() +
  theme(legend.position = "bottom") +
  ylab("") +
  xlab("") +
  NULL
g
ggsave(g, file = "../Figures/total_detected.pdf", height = 6, width = 10)



boundary_polygons$total <- colData(sce)[boundary_polygons$uniqueID, "total"]
g = ggplot(data = subset(boundary_polygons, 
                         z == 2 & embryo %in% c("embryo1","embryo2"))) + 
  geom_polygon(aes(x = segmentation_vertices_x_global_affine,
                   y = -segmentation_vertices_y_global_affine,
                   group = uniqueID,
                   fill = total),
               colour = "black", size = 0.1, alpha = 0.7) +
  theme_transparent() +
  theme_classic() +
  coord_fixed() +
  facet_wrap(~embryo) +
  theme(legend.position = "none") +
  scale_colour_gradient2(low = "gray75", high = "cornflowerblue") +
  theme(axis.line = element_blank()) + 
  theme(axis.text = element_blank()) + 
  theme(axis.title = element_blank()) +
  theme(axis.ticks = element_blank()) +
  NULL
g
```

# Joint UMAP with seqFISH and mouse gastrulation atlas

```{r}
joint_df = readRDS("../analysis_output/E8.5/old/E8.5_joint_mnn_df.Rds")

joint_df_sub = subset(joint_df, type == "atlas" | z %in% c(2,5))

g = ggplot(joint_df_sub[sample(nrow(joint_df_sub)),], aes(x = UMAP1, y = UMAP2)) + 
  geom_point(aes(colour = type), size = 0.5, alpha = 0.2) + 
  theme_classic() + 
  scale_colour_manual(values = c(atlas = "orange", spatial = "navy")) +
  theme(legend.position = "bottom") +
  theme(legend.text = element_text(size = 15)) +
  ggtitle("Joint UMAP") +
  coord_fixed() +
  xlab("UMAP1") +
  ylab("UMAP2") +
  theme(axis.text = element_blank()) +
  theme(axis.ticks = element_blank()) +
  guides(colour = guide_legend(title = "",
                               override.aes = list(size = 15, alpha = 1))) +
  NULL
g

ggsave(g, file = "../Figures/joint_umap.pdf", height = 6, width = 5)

joint_df_sub$celltype_atlas = joint_df_sub$celltype_mapped
joint_df_sub$celltype_atlas[joint_df_sub$type == "spatial"] <- NA

g = ggplot(joint_df_sub, aes(x = UMAP1, y = UMAP2)) + 
  geom_point(aes(colour = celltype_atlas, alpha = type), size = 0.5) + 
  theme_classic() + 
  scale_colour_manual(values = celltype_colours, na.value = "navy") +
  theme(legend.position = "none") +
  theme(legend.text = element_text(size = 15)) +
  scale_alpha_manual(values = c("atlas" = 1, "spatial" = 0.2)) +
  ggtitle("Joint UMAP") +
  coord_fixed() +
  xlab("UMAP1") +
  ylab("UMAP2") +
  theme(axis.text = element_blank()) +
  theme(axis.ticks = element_blank()) +
  guides(colour = guide_legend(title = "",
                               override.aes = list(size = 15, alpha = 1))) +
  NULL
g

ggsave(g, file = "../Figures/joint_umap_celltypes.pdf", height = 5, width = 5)
```

# celltypes legend

```{r}
ctypes_detected = names(celltype_colours)[which(names(celltype_colours) %in% sce$celltype_mapped[sce$z %in% c(2,5)])]
ctypes_df = data.frame( x = seq_len(length(ctypes_detected)),
                        ctype = ctypes_detected)

celltype_clean_labels = names(celltype_colours)
names(celltype_clean_labels) <- names(celltype_colours)
celltype_clean_labels["Haematoendothelial progenitors"] <- "HEP"
celltype_clean_labels["Gut"] <- "Gut tube"
celltype_clean_labels["ExE mesoderm"] <- "Extraembryonic mesoderm"
celltype_clean_labels["Def. endoderm"] <- "Definitive endoderm"


g = ggplot(ctypes_df, aes(x = x, y = x)) + 
  geom_point(aes(colour = ctype), size = 10) + 
  theme_classic() + 
  theme_transparent() +
  scale_colour_manual(values = celltype_colours, 
                      labels = celltype_clean_labels) + 
  theme(legend.position = "right") +
  guides(colour = guide_legend(ncol = 1, title = "")) + 
  theme(legend.text = element_text(size = 18)) +
  NULL

gg = as_ggplot(get_legend(g))
gg

ctypes_df$order = -rank(as.character(ctypes_df$ctype))

g = ggplot(ctypes_df, aes(y = order)) + 
  geom_point(aes(colour = ctype), x = 0, size = 5) + 
  geom_text(aes(label = ctype), x = 0.05, hjust = 0) +
  theme_classic() +
  theme(legend.position = "none") +
  theme(axis.text = element_blank()) + 
  theme(axis.line = element_blank()) +
  theme(axis.ticks = element_blank()) +
  xlab("") + 
  ylab("") +
  scale_colour_manual(values = celltype_colours) +
  NULL

ggsave(g, file = "../Figures/celltypes_legend.pdf", height = 8, width = 3)
```

# Celltype proportions for AP pseudo

```{r}
now_joint_now = subset(now_joint_df, split == "Nowotschin")
now_joint_sce = subset(now_joint_df, split != "Nowotschin")
now_joint_sce1 = subset(now_joint_df, split %in% c("embryo12", "embryo15"))
now_joint_sce2 = subset(now_joint_df, split %in% c("embryo22", "embryo25"))
now_joint_sce3 = subset(now_joint_df, split %in% c("embryo32", "embryo35"))
# now_joint_now = now_joint_df

ctypes_all = sort(unique(now_joint_now$ctype))

celltype_props = data.frame(
  celltype = ctypes_all,
  atlas_all = unclass(table(factor(now_joint_now$ctype, levels = ctypes_all))),
  spatial_all = unclass(table(factor(now_joint_sce$ctype, levels = ctypes_all))),
  spatial_embryo1 = unclass(table(factor(now_joint_sce1$ctype, levels = ctypes_all))),
  spatial_embryo2 = unclass(table(factor(now_joint_sce2$ctype, levels = ctypes_all))),
  spatial_embryo3 = unclass(table(factor(now_joint_sce3$ctype, levels = ctypes_all)))
)

plot(1+celltype_props$atlas_all, 1+celltype_props$spatial_all, log = "xy")

plot(1+celltype_props$atlas, 1+celltype_props$spatial_all, log = "xy")
plot(1+celltype_props$atlas, 1+celltype_props$spatial_embryo1, log = "xy")


g1 = ggplot(celltype_props, aes(x = 1 + atlas_all, y = 1 + spatial_embryo1)) + 
  geom_point() + 
  geom_text_repel(aes(label = celltype)) + 
  theme_classic() +
  scale_x_log10() + 
  scale_y_log10() +
  geom_smooth(method = "lm") +
  xlab("Number of cells in E8.5 atlas") +
  ylab("Number of cells in E8.5 embryo 1") +
  NULL

g2 = ggplot(celltype_props, aes(x = 1 + atlas_all, y = 1 + spatial_embryo2)) + 
  geom_point() + 
  geom_text_repel(aes(label = celltype)) + 
  theme_classic() +
  scale_x_log10() + 
  scale_y_log10() +
  geom_smooth(method = "lm") +
  xlab("Number of cells in E8.5 atlas") +
  ylab("Number of cells in E8.5 embryo 2") +
  NULL

g3 = ggplot(celltype_props, aes(x = 1 + atlas_all, y = 1 + spatial_embryo3)) + 
  geom_point() + 
  geom_text_repel(aes(label = celltype)) + 
  theme_classic() +
  scale_x_log10() + 
  scale_y_log10() +
  geom_smooth(method = "lm") +
  xlab("Number of cells in E8.5 atlas") +
  ylab("Number of cells in E8.5 embryo 3") +
  NULL

g1 + g2 + g3

g = ggplot(celltype_props, aes(x = 1 + atlas_all)) +
  geom_point(aes(y = 1 + spatial_embryo1), colour = "orange") + 
  geom_point(aes(y = 1 + spatial_embryo2), colour = "red") + 
  geom_point(aes(y = 1 + spatial_embryo3), colour = "blue") + 
  theme_classic() +
  scale_x_log10() + 
  scale_y_log10() +
  geom_smooth(method = "lm", aes(y = 1 + spatial_embryo1)) +
  geom_smooth(method = "lm", aes(y = 1 + spatial_embryo2)) +
  geom_smooth(method = "lm", aes(y = 1 + spatial_embryo3)) +
  xlab("Number of cells in E8.5 atlas") +
  ylab("Number of cells in E8.5 embryo 3") +
  NULL
g

lm1 = lm(I(log10(1+spatial_embryo1)) ~ I(log10(1+atlas_all)), data = celltype_props)
celltype_props$lm1_res <- lm1$residuals

lm2 = lm(I(log10(1+spatial_embryo2)) ~ I(log10(1+atlas_all)), data = celltype_props)
celltype_props$lm2_res <- lm2$residuals

lm3 = lm(I(log10(1+spatial_embryo3)) ~ I(log10(1+atlas_all)), data = celltype_props)
celltype_props$lm3_res <- lm3$residuals

lm_df = data.frame(
  celltype = c(names(lm1$residuals),names(lm2$residuals),names(lm3$residuals)),
  residual = c(lm1$residuals,lm2$residuals,lm3$residuals),
  embryo = rep(c("embryo1","embryo2","embryo3"), 
               times = c(length(lm1$residuals),length(lm2$residuals),length(lm3$residuals)))
)

lm_df$celltype <- factor(lm_df$celltype, levels = rev(c(names(sort(tapply(lm_df$residual, lm_df$celltype, mean))))))

g = ggplot(lm_df, aes(x = celltype, group = embryo, fill = celltype, y = residual)) + 
  geom_hline(yintercept = c(-1,1)*0.5, linetype = "dashed", colour = "grey") +
  geom_col(position = "dodge", colour = "black") + 
  theme_classic() + 
  theme(axis.text.x = element_text(angle = 60, size = 15, hjust = 1)) +
  theme(axis.title.y = element_text(size = 15)) +
  xlab("") +
  theme(legend.position = "none") +
  scale_alpha_manual(values = c("TRUE" = 1, "FALSE" = 0.3)) +
  scale_fill_manual(values = AP_pseudo_colours) +
  ylab("Relative enrichment") +
  NULL
g

ggsave(g, file = "../Figures/AP_pseudo_enrichment.pdf", height = 4, width = 8)
```

# smFISH boxplot of field of view spots detected

```{r}
sm = readRDS("../analysis_output/smFISH_counts.Rds")

# calculate the number of dots divided by the total seqFISH dots
cells = intersect(colnames(sm), colnames(sce))
table(colnames(sm) %in% cells)
table(colnames(sce) %in% cells)

# proportion of the seqFISH counts
sm_int = t(t(sm[,cells]) / sce[,cells]$total) 

sm_int_long = melt(sm_int)
colnames(sm_int_long) <- c("Gene", "uniqueID", "exp")
sm_int_long$embryo_pos_z <- colData(sce)[as.character(sm_int_long$uniqueID), "embryo_pos_z"]
sm_int_long$z <- colData(sce)[as.character(sm_int_long$uniqueID), "z"]
sm_int_long$embryo <- colData(sce)[as.character(sm_int_long$uniqueID), "embryo"]

sm_int_long_split = split.data.frame(sm_int_long, sm_int_long$z)

sm_int_long_2 = sm_int_long_split[["2"]]
sm_int_long_2$embryo_pos_z <- droplevels(sm_int_long_2$embryo_pos_z)
sm_int_long_5 = sm_int_long_split[["5"]]
sm_int_long_5$embryo_pos_z <- droplevels(sm_int_long_5$embryo_pos_z)

sm_int_long_2$exp_rel = unsplit(tapply(sm_int_long_2$exp, sm_int_long_2$Gene, function(x) x/max(x)),
                                sm_int_long_2$Gene)

sm_int_long_5$exp_rel = unsplit(tapply(sm_int_long_5$exp, sm_int_long_5$Gene, function(x) x/max(x)),
                                sm_int_long_5$Gene)

sm_int_long_summary_2 = tapply(sm_int_long_2$exp_rel, list(sm_int_long_2$embryo_pos_z,
                                                           sm_int_long_2$Gene),
                               mean)

g = ggplot(sm_int_long[1:50,], 
           aes(x = embryo_pos_z,
               y = exp)) + 
  theme_classic() + 
  geom_violin() + 
  facet_grid(Gene ~ .) + 
  NULL

ggsave(g, file = "../Figures/smFISH_proportion_fov.pdf",
       height = 20, width = 15)
```

# alternative use cell neighbourhood to assess field of view effect for smFISH

```{r}
graph = readRDS("../analysis_output/E8.5/E8.5_neighbourGraph_1.3.Rds")

dim(sm)

sm_raw = readRDS("../analysis_output/smFISH_counts.Rds")

# calculate the number of dots divided by the total seqFISH dots
cells = intersect(colnames(sm_raw), colnames(sce))
table(colnames(sm_raw) %in% cells)
table(colnames(sce) %in% cells)

sm <- log10(1+sm_raw[,cells])
pos_sm = as.character(sce[,cells]$embryo_pos)
names(pos_sm) <- cells

egraph = get.edgelist(graph)
rownames(egraph) <- paste0("pair_",seq_len(nrow(egraph))) 
graph_pos = cbind(pos_sm[egraph[,1]], pos_sm[egraph[,2]])
rownames(graph_pos) <- rownames(egraph)
graph_pos_sub = na.omit(graph_pos[graph_pos[,1] != graph_pos[,2],])

out_high = matrix(NA,
                  ncol = length(sort(unique(pos_sm))),
                  nrow = length(rownames(sm)),
                  dimnames = list(rownames(sm),
                                  sort(unique(pos_sm))))

out_low = matrix(NA,
                 ncol = length(sort(unique(pos_sm))),
                 nrow = length(rownames(sm)),
                 dimnames = list(rownames(sm),
                                 sort(unique(pos_sm))))

for (fov in sort(unique(pos_sm))) {
  
  print(fov)
  
  graph_pos_sub_sub = graph_pos_sub[graph_pos_sub[,1] == fov | graph_pos_sub[,2] == fov, ]
  pairs_sub = egraph[rownames(graph_pos_sub_sub), ]
  cells_sub = sort(unique(c(pairs_sub)))
  
  for (gene in rownames(sm)) {
    
    # print(gene)
    
    # calculate mean difference in expression of the gene for these cells
    sm_exp_sub = sm[gene, cells_sub]
    sm_exp_mean = tapply(sm_exp_sub, pos_sm[cells_sub], mean)
    out_high[gene, fov] <- max(sm_exp_mean[fov] - sm_exp_mean[setdiff(names(sm_exp_mean), fov)])
    
    out_low[gene, fov] <- min(sm_exp_mean[fov] - sm_exp_mean[setdiff(names(sm_exp_mean), fov)])
    
  }
}


h1 = Heatmap(out_high,
             col = c("grey", "purple", "black"),
             name = "Maximum mean\ndifference in\nneighbouring\nfield of view",
             use_raster = TRUE
)
h1

pdf(file = "../Figures/smFISH_neighbour_comparison.pdf", height = 10, width = 20)
print(h1)
dev.off()

h2 = Heatmap(t(out_low),
             col = rev(c("grey", "purple", "black")))

h1 + h2
```

# Finish

```{r}
sessionInfo()
```